{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "15547c3c",
   "metadata": {},
   "source": [
    "## Convert files from Thermo\n",
    "\n",
    "ThermoRawFileParser https://github.com/compomics/ThermoRawFileParser\n",
    "\n",
    "Convert profile MS1 data from Thermo raw files to centroid (MS1 and MS2) mzml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f30ec3b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-10-21 16:40:05 INFO Started analyzing folder data/raw/\n",
      "2021-10-21 16:40:05 INFO The folder contains 6 RAW files\n",
      "2021-10-21 16:40:05 INFO Started parsing data/raw/FilipinIII.raw\n",
      "2021-10-21 16:40:06 INFO Processing 3890 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-10-21 16:40:15 INFO Finished parsing data/raw/FilipinIII.raw\n",
      "2021-10-21 16:40:15 INFO Started parsing data/raw/Pentamycin.raw\n",
      "2021-10-21 16:40:17 INFO Processing 3827 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-10-21 16:40:25 INFO Finished parsing data/raw/Pentamycin.raw\n",
      "2021-10-21 16:40:25 INFO Started parsing data/raw/Leupeptin.raw\n",
      "2021-10-21 16:40:26 INFO Processing 3907 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-10-21 16:40:34 INFO Finished parsing data/raw/Leupeptin.raw\n",
      "2021-10-21 16:40:34 INFO Started parsing data/raw/Epemicins.raw\n",
      "2021-10-21 16:40:35 INFO Processing 4611 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-10-21 16:40:45 INFO Finished parsing data/raw/Epemicins.raw\n",
      "2021-10-21 16:40:45 INFO Started parsing data/raw/GermicidinB.raw\n",
      "2021-10-21 16:40:45 INFO Processing 3619 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-10-21 16:40:53 INFO Finished parsing data/raw/GermicidinB.raw\n",
      "2021-10-21 16:40:53 INFO Started parsing data/raw/GermicidinA.raw\n",
      "2021-10-21 16:40:54 INFO Processing 3625 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-10-21 16:40:59 INFO Finished parsing data/raw/GermicidinA.raw\n"
     ]
    }
   ],
   "source": [
    "!mono resources/ThermoRawFileParser/ThermoRawFileParser.exe -d=\"data/raw/\" -o=\"results/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82179e30",
   "metadata": {},
   "source": [
    "## Copy all the files to data/raw directory and you can use the following script to generate a samples.tsv file automatically:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cc98040f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>comment</th>\n",
       "      <th>MAPnumber</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Epemicins</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GermicidinA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GermicidinB</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name comment MAPnumber\n",
       "0    Epemicins                  \n",
       "1  GermicidinA                  \n",
       "2  GermicidinB                  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!(cd data/raw && ls *.raw > filelist.txt)\n",
    "import pandas as pd\n",
    "header_list = [\"sample_name\"]\n",
    "df=pd.read_csv(\"data/raw/filelist.txt\", names=header_list)\n",
    "df[\"sample_name\"]=df[\"sample_name\"].replace(\".raw\", value=\"\", regex=True)\n",
    "df[\"comment\"] = \"\"\n",
    "df[\"MAPnumber\"] = \"\"\n",
    "df.to_csv(\"config/samples.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7c0f75",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "969483bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Determination of memory status is not supported on this \n",
      " platform, measuring for memoryleaks will never fail\n"
     ]
    }
   ],
   "source": [
    "from pyopenms import *\n",
    "from pandas import DataFrame\n",
    "import pandas as pd\n",
    "import pyteomics\n",
    "from pyteomics.openms import featurexml\n",
    "import numpy as np\n",
    "from pyteomics import mztab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da76ec1a",
   "metadata": {},
   "source": [
    "# Preprocessing step\n",
    "\n",
    "## The first preprocessing function consists of six steps:\n",
    "\n",
    "### 1) PrecursorCorrection (To the \"highest intensity MS1 peak\")\n",
    "\n",
    "This algorithm is used directly after the file introduction, in order to correct any wrong MS1 precursor annotation - this means it will correct the precursos m/z in a given mass range (e.g. 0.1 Da) and then use the highest intensity peak as the precursor. We assume that, in the given mass window, the precursor with the hightest intensity was actually fragmented (top-n method), which is a method used by default in the Thermo Orbitrap instrument (Center for Biosustainability).\n",
    "\n",
    "### OpenMS definition:\n",
    "\n",
    "\"Selection of the peak with the highest intensity as corrected precursor mass in a given mass range (e.g. precursor mass +/- 0.2 Da)\n",
    "\n",
    "For each MS2 spectrum the corresponding MS1 spectrum is determined by using the rt information of the precursor. In the MS1, the peak with the highest intensity in a given mass range to the uncorrected precursor m/z is selected and used as corrected precursor m/z.\"\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html//classOpenMS_1_1PrecursorCorrection.html#a8acf85ba8b9f249de0369bb083355982 \n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b538e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_filename= \"results/Epemicins/interim/Epemicins.mzML\" #introduce the mzML file that you want to preprocess here. If it is more than one, refer to the other jupyter notebook\n",
    "exp = MSExperiment()\n",
    "MzMLFile().load(example_filename, exp)\n",
    "exp.sortSpectra(True)\n",
    "    \n",
    "delta_mzs= []\n",
    "mzs = []\n",
    "rts= []\n",
    "PrecursorCorrection.correctToHighestIntensityMS1Peak(exp, 100.0, True, delta_mzs, mzs, rts)\n",
    "MzMLFile().store(\"results/Epemicins/interim/precursorcorrected_Epemicins.mzML\", exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16c4b46",
   "metadata": {},
   "source": [
    "### 2) Mass trace detection\n",
    "\n",
    "A mass trace extraction method that gathers peaks similar in m/z and moving along retention time.\n",
    "\n",
    "Peaks of a MSExperiment are sorted by their intensity and stored in a list of potential chromatographic apex positions. Only peaks that are **above the noise threshold** (user-defined) are analyzed and only peaks that are n times above this minimal threshold are considered as apices. This saves computational resources and decreases the noise in the resulting output.\n",
    "\n",
    "Starting with these, mass traces are extended in retention time. During this extension phase, the centroid m/z is computed on-line as an intensity-weighted mean of peaks.\n",
    "\n",
    "The extension phase ends when either the frequency of gathered peaks drops below a threshold (min_sample_rate, see MassTraceDetection parameters) or when the number of missed scans exceeds a threshold (trace_termination_outliers, see MassTraceDetection parameters).\n",
    "\n",
    "Finally, only mass traces that pass a filter (a certain minimal and maximal length as well as having the minimal sample rate criterion fulfilled) get added to the result.\"\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1MassTraceDetection.html#abff6e392ce6da7af8f083397494a7971\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5f5c9697",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'mass trace detection':\n",
      "-- done [took 0.44 s (CPU), 0.43 s (Wall)] -- \n"
     ]
    }
   ],
   "source": [
    "mass_traces = []\n",
    "mtd = MassTraceDetection()\n",
    "mtd_par = mtd.getDefaults()\n",
    "mtd_par.setValue(\"mass_error_ppm\", 10.0) \n",
    "mtd_par.setValue(\"noise_threshold_int\", 1.0e04)\n",
    "mtd.setParameters(mtd_par)\n",
    "mtd.run(exp, mass_traces, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561c3ab9",
   "metadata": {},
   "source": [
    "### 3) Elution peak detection\n",
    "\n",
    "\"Extracts chromatographic peaks from a mass trace.\n",
    "\n",
    "Mass traces may consist of several consecutively (partly overlapping) eluting peaks, e.g., stemming from (almost) isobaric compounds that are separated by retention time. Especially in metabolomics, isomeric compounds with exactly the same mass but different retentional behaviour may still be contained in the same mass trace.\n",
    "\n",
    "This method first applies smoothing on the mass trace's intensities, then detects local minima/maxima in order to separate the chromatographic peaks from each other. Detection of maxima is performed on the smoothed intensities and uses a fixed peak width (given as parameter chrom_fwhm) within which only a single maximum is expected. Currently smoothing is done using SavitzkyGolay smoothing with a second order polynomial and a frame length of the fixed peak width.\n",
    "\n",
    "Depending on the \"width_filtering\" parameters, mass traces are filtered by length in seconds (\"fixed\" filter) or by quantile.\n",
    "\n",
    "The output of the algorithm is a set of chromatographic peaks for each mass trace, i.e. a vector of split mass traces (see ElutionPeakDetection parameters).\n",
    "\n",
    "In general, a user would want to call the \"detectPeaks\" functions, potentially followed by the \"filterByPeakWidth\" function.\"\n",
    "\n",
    "This method in other words is \"deconvolution\".\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1ElutionPeakDetection.html#details\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "37d0c726",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'elution peak detection':\n",
      "-- done [took 2.18 s (CPU), 0.31 s (Wall)] -- \n"
     ]
    }
   ],
   "source": [
    "mass_traces_split = []\n",
    "mass_traces_final = []\n",
    "epd = ElutionPeakDetection()\n",
    "epd_par = epd.getDefaults()\n",
    "epd_par.setValue(\"width_filtering\", \"fixed\")\n",
    "epd.setParameters(epd_par)\n",
    "epd.detectPeaks(mass_traces, mass_traces_split)\n",
    "    \n",
    "if (epd.getParameters().getValue(\"width_filtering\") == \"auto\"):\n",
    "     epd.filterByPeakWidth(mass_traces_split, mass_traces_final)\n",
    "else:\n",
    "    mass_traces_final = mass_traces_split"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e1dcab",
   "metadata": {},
   "source": [
    "### 4) Feature detection\n",
    "\n",
    "FeatureFinderMetabo assembles metabolite features from singleton mass traces.\n",
    "\n",
    "Mass traces alone would allow for further analysis such as metabolite ID or statistical evaluation. However, in general, monoisotopic mass traces are accompanied by satellite C13 peaks and thus may render the analysis more difficult. **FeatureFinderMetabo fulfills a further data reduction step by assembling compatible mass traces to metabolite features** (that is, all mass traces originating from one metabolite). To this end, multiple metabolite hypotheses are formulated and scored according to how well differences in RT (optional), m/z or intensity ratios match to those of theoretical isotope patterns.\n",
    "\n",
    "Detailed information is stored in the featureXML output: it contains meta-values for each feature about the mass trace differences (inspectable via TOPPView). \n",
    "\n",
    "By default, the linear model is used (better for small molecules - use the 13C for proteomics).\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_FeatureFinderMetabo.html\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c31f3921",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'assembling mass traces to features':\n",
      "-- done [took 4.40 s (CPU), 0.71 s (Wall)] -- \n"
     ]
    }
   ],
   "source": [
    "feature_map_FFM = FeatureMap()\n",
    "feat_chrom = []\n",
    "ffm = FeatureFindingMetabo()\n",
    "ffm_par = ffm.getDefaults() \n",
    "ffm_par.setValue(\"isotope_filtering_model\", \"none\")\n",
    "ffm_par.setValue(\"remove_single_traces\", \"false\")\n",
    "ffm_par.setValue(\"mz_scoring_by_elements\", \"false\")\n",
    "ffm_par.setValue(\"report_convex_hulls\", \"true\")\n",
    "ffm.setParameters(ffm_par)\n",
    "ffm.run(mass_traces_final, feature_map_FFM, feat_chrom)\n",
    "feature_map_FFM.setUniqueIds()\n",
    "fh = FeatureXMLFile()\n",
    "filename= \"results/Epemicins/interim/precursorcorrected_Epemicins.mzML\"\n",
    "feature_map_FFM.setPrimaryMSRunPath([filename.encode()])\n",
    "fh.store('./results/Epemicins/interim/FFM_Epemicins.featureXML', feature_map_FFM)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50d5ee28",
   "metadata": {},
   "source": [
    "### 5) Metabolite adduct decharger (MetaboliteFeatureDeconvolution)\n",
    "\n",
    "For each peak, this algorithm reconstructs neutral masses by enumerating all possible adducts with matching charge. Provide the probability of adduct formation according to your instrument using the parameter \"potential adducts\".\n",
    "\n",
    "With SIRIUS, you are only able to use singly charged adducts so charges higher than 1 are filtered out.\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/UTILS_MetaboliteAdductDecharger.html\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "59a21b2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MassExplainer table size: 12\n",
      "4611 spectra and 1 chromatograms stored.\n",
      "Generating Masses with threshold: -6.90776 ...\n",
      "done\n",
      "0 of 2852 valid net charge compomer results did not pass the feature charge constraints\n",
      "Inferring edges raised edge count from 2852 to 2852\n",
      "Found 2852 putative edges (of 131864) and avg hit-size of 1\n",
      "Using solver 'coinor' ...\n",
      "Optimal solution found!\n",
      "<Using solver 'coinor' ...> occurred 7 times\n",
      " Branch and cut took 0.919725 seconds,  with objective value: 6.88.\n",
      "<Optimal solution found!> occurred 7 times\n",
      "ILP score is: 6.88\n",
      "Agreeing charges: 541/3526\n",
      "ConsensusFeature::computeDechargeConsensus() WARNING: Feature's charge is 0! This will lead to M=0!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Compomer: Da -41.0003; q_net 0; logP -3.91202[[ (H2O1Na1) --> () ]]\n",
      " Compomer: Da -36.0449; q_net 0; logP -4.60517[[ (H2O1H4N1) --> () ]]\n",
      " Compomer: Da -21.9819; q_net 0; logP -2.12026[[ (Na1) --> (H1) ]]\n",
      " Compomer: Da -19.0184; q_net 0; logP -2.81341[[ (H1H2O1) --> () ]]\n",
      " Compomer: Da -17.0265; q_net 0; logP -2.81341[[ (H4N1) --> (H1) ]]\n",
      " Compomer: Da -4.9554; q_net 0; logP -3.91202[[ (Na1) --> (H4N1) ]]\n",
      " Compomer: Da 4.9554; q_net 0; logP -3.91202[[ (H4N1) --> (Na1) ]]\n",
      " Compomer: Da 17.0265; q_net 0; logP -2.81341[[ (H1) --> (H4N1) ]]\n",
      " Compomer: Da 19.0184; q_net 0; logP -2.81341[[ () --> (H1H2O1) ]]\n",
      " Compomer: Da 21.9819; q_net 0; logP -2.12026[[ (H1) --> (Na1) ]]\n",
      " Compomer: Da 36.0449; q_net 0; logP -4.60517[[ () --> (H2O1H4N1) ]]\n",
      " Compomer: Da 41.0003; q_net 0; logP -3.91202[[ () --> (H2O1Na1) ]]\n",
      " "
     ]
    }
   ],
   "source": [
    "mfd = MetaboliteFeatureDeconvolution()\n",
    "mdf_par = mfd.getDefaults()\n",
    "mdf_par.setValue(\"potential_adducts\",  [b\"H:+:0.6\",b\"Na:+:0.2\",b\"NH4:+:0.1\", b\"H2O:-:0.1\"])\n",
    "mdf_par.setValue(\"charge_min\", 1, \"Minimal possible charge\")\n",
    "mdf_par.setValue(\"charge_max\", 1, \"Maximal possible charge\")\n",
    "mdf_par.setValue(\"charge_span_max\", 1)\n",
    "mdf_par.setValue(\"max_neutrals\", 1)\n",
    "mfd.setParameters(mdf_par)\n",
    "    \n",
    "feature_map_DEC = FeatureMap()\n",
    "cons_map0 = ConsensusMap()\n",
    "cons_map1 = ConsensusMap()\n",
    "mfd.compute(feature_map_FFM, feature_map_DEC, cons_map0, cons_map1)\n",
    "fxml = FeatureXMLFile()\n",
    "fh.store('./results/Epemicins/interim/MFD_Epemicins.featureXML', feature_map_DEC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99c2993c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from functools import reduce\n",
    "from pathlib import Path\n",
    "from time import perf_counter\n",
    "import sys\n",
    "\n",
    "from IPython.core.display import display\n",
    "from pandas import CategoricalDtype\n",
    "import numpy as np\n",
    "from pyopenms import *\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "class FeatureMapDF(FeatureMap):      \n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "\n",
    "    def get_df(self):\n",
    "        def gen(fmap: FeatureMap, fun):\n",
    "            for f in fmap:\n",
    "                yield from fun(f)\n",
    "\n",
    "        def extractMetaData(f: Feature):\n",
    "            metavals = []\n",
    "            types = []\n",
    "            for k in metavals:\n",
    "                if k == b\"num_of_masstraces\":\n",
    "                    types.append('f')\n",
    "                elif k== b\"FWHM\":\n",
    "                    types.append('f')\n",
    "                elif k== b\"max_height\":\n",
    "                    types.append('i4')\n",
    "                elif k== b\"masstrace_intensity\":\n",
    "                    types.append(\"f\")\n",
    "            # subfeatures = f.getFeatureList()  # type: list[FeatureHandle]\n",
    "            pep = f.getPeptideIdentifications()  # type: list[PeptideIdentification]\n",
    "            bb = f.getConvexHull().getBoundingBox2D()\n",
    "            for u in metavals:\n",
    "                f.getMetaValue(u)\n",
    "            if len(pep) != 0:\n",
    "                hits = pep[0].getHits()\n",
    "                if len(hits) != 0:\n",
    "                    besthit = hits[0]  # type: PeptideHit\n",
    "                    yield f.getUniqueId(), besthit.getSequence().toString(), f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"num_of_masstraces\"), f.getOverallQuality(), f.getIntensity()\n",
    "                else:\n",
    "                    yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"num_of_masstraces\"), f.getOverallQuality(), f.getIntensity()\n",
    "            else:\n",
    "                yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"num_of_masstraces\"), f.getOverallQuality(), f.getIntensity()\n",
    "\n",
    "        cnt = self.size()\n",
    "\n",
    "        mddtypes = [('id', np.dtype('uint64')), ('sequence', 'U200'), ('charge', 'i4'), ('RT', 'f'), ('mz', 'f'),\n",
    "                    ('RTstart', 'f'), ('RTend', 'f'), (\"num_of_masstraces\", 'f'),\n",
    "                    ('quality', 'f'), ('intensity', 'f')]\n",
    "        mdarr = np.fromiter(iter=gen(self, extractMetaData), dtype=mddtypes, count=cnt)\n",
    "        df= pd.DataFrame(mdarr).set_index('id').sort_values(\"mz\").drop(columns= \"sequence\")\n",
    "        df= df[df[\"num_of_masstraces\"]>=2]\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50aeae13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>charge</th>\n",
       "      <th>RT</th>\n",
       "      <th>mz</th>\n",
       "      <th>RTstart</th>\n",
       "      <th>RTend</th>\n",
       "      <th>num_of_masstraces</th>\n",
       "      <th>quality</th>\n",
       "      <th>intensity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15695211054020622558</th>\n",
       "      <td>1</td>\n",
       "      <td>329.607452</td>\n",
       "      <td>208.133240</td>\n",
       "      <td>326.792847</td>\n",
       "      <td>335.102264</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>3.212667e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11553075680875761771</th>\n",
       "      <td>1</td>\n",
       "      <td>310.614807</td>\n",
       "      <td>208.133255</td>\n",
       "      <td>303.933105</td>\n",
       "      <td>315.405701</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>1.985221e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16931706070924112617</th>\n",
       "      <td>1</td>\n",
       "      <td>379.585754</td>\n",
       "      <td>260.185669</td>\n",
       "      <td>362.698792</td>\n",
       "      <td>455.897400</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>4.589339e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464053971863077264</th>\n",
       "      <td>1</td>\n",
       "      <td>228.511551</td>\n",
       "      <td>261.123413</td>\n",
       "      <td>222.785919</td>\n",
       "      <td>235.216217</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>4.301382e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14207239120303140199</th>\n",
       "      <td>1</td>\n",
       "      <td>316.367584</td>\n",
       "      <td>271.060028</td>\n",
       "      <td>302.927643</td>\n",
       "      <td>321.092896</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>1.677704e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13812678751659148901</th>\n",
       "      <td>1</td>\n",
       "      <td>350.433289</td>\n",
       "      <td>1603.895020</td>\n",
       "      <td>348.765778</td>\n",
       "      <td>355.269623</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.001192</td>\n",
       "      <td>1.190649e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4289608483695663489</th>\n",
       "      <td>1</td>\n",
       "      <td>342.791321</td>\n",
       "      <td>1604.396851</td>\n",
       "      <td>337.522614</td>\n",
       "      <td>352.043243</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>4.679359e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7662137660621363607</th>\n",
       "      <td>2</td>\n",
       "      <td>47.941818</td>\n",
       "      <td>1639.049805</td>\n",
       "      <td>43.264481</td>\n",
       "      <td>61.428474</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>2.294459e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7650966078180970882</th>\n",
       "      <td>1</td>\n",
       "      <td>173.429962</td>\n",
       "      <td>1646.474365</td>\n",
       "      <td>169.596756</td>\n",
       "      <td>181.128769</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.192035e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12543873065339080715</th>\n",
       "      <td>2</td>\n",
       "      <td>105.360298</td>\n",
       "      <td>1775.552734</td>\n",
       "      <td>101.564934</td>\n",
       "      <td>115.742714</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>1.904661e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2428 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      charge          RT           mz     RTstart       RTend  \\\n",
       "id                                                                              \n",
       "15695211054020622558       1  329.607452   208.133240  326.792847  335.102264   \n",
       "11553075680875761771       1  310.614807   208.133255  303.933105  315.405701   \n",
       "16931706070924112617       1  379.585754   260.185669  362.698792  455.897400   \n",
       "2464053971863077264        1  228.511551   261.123413  222.785919  235.216217   \n",
       "14207239120303140199       1  316.367584   271.060028  302.927643  321.092896   \n",
       "...                      ...         ...          ...         ...         ...   \n",
       "13812678751659148901       1  350.433289  1603.895020  348.765778  355.269623   \n",
       "4289608483695663489        1  342.791321  1604.396851  337.522614  352.043243   \n",
       "7662137660621363607        2   47.941818  1639.049805   43.264481   61.428474   \n",
       "7650966078180970882        1  173.429962  1646.474365  169.596756  181.128769   \n",
       "12543873065339080715       2  105.360298  1775.552734  101.564934  115.742714   \n",
       "\n",
       "                      num_of_masstraces   quality     intensity  \n",
       "id                                                               \n",
       "15695211054020622558                2.0  0.000180  3.212667e+07  \n",
       "11553075680875761771                2.0  0.000011  1.985221e+06  \n",
       "16931706070924112617                2.0  0.000258  4.589339e+07  \n",
       "2464053971863077264                 2.0  0.000025  4.301382e+06  \n",
       "14207239120303140199                2.0  0.000009  1.677704e+06  \n",
       "...                                 ...       ...           ...  \n",
       "13812678751659148901                3.0  0.001192  1.190649e+08  \n",
       "4289608483695663489                 2.0  0.000371  4.679359e+07  \n",
       "7662137660621363607                 2.0  0.000003  2.294459e+05  \n",
       "7650966078180970882                 3.0  0.000006  5.192035e+05  \n",
       "12543873065339080715                6.0  0.000049  1.904661e+06  \n",
       "\n",
       "[2428 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filename= './results/Epemicins/interim/MFD_Epemicins.featureXML'\n",
    "fmap = FeatureMapDF()\n",
    "FeatureXMLFile().load(filename, fmap)\n",
    "DF= fmap.get_df()\n",
    "feature_csv= './results/Epemicins/features_Epemicins.csv'\n",
    "DF.to_csv(feature_csv)\n",
    "display(DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14b8a4ae",
   "metadata": {},
   "source": [
    "### 6) PrecursorCorrection (To the \"nearest feature”)\n",
    "\n",
    "This algorithm is used after feature detection, adduct grouping and even identification via accurate mass search. It basically allows the precursor correction on MS2 level. \n",
    "\n",
    "Which means that if there are MS2 spectra in my feature space which have been measured in isotope traces, it “corrects” the MS2 spectrum annotation to the monoisotopic trace. That is why you have a high mass deviation 100 pm, but 0.0 rt tolerance. So it basically corrects the MS2 to the feature centroid that can be found/mapped by SIRIUS::preprocessing later on.\n",
    "\n",
    "Wrong assignment of the mono-isotopic mass for precursors are assumed:\n",
    "\n",
    "1. if precursor_mz matches the mz of a non-monoisotopic feature mass trace and \n",
    "2. in the case that believe_charge is true: if feature_charge matches the precursor_charge In the case of wrong mono-isotopic assignment several options for correction are available: keep_original will create a copy of the precursor and tandem spectrum for the new mono-isotopic mass trace and retain the original one. all_matching_features does this not for only the closest feature but all features in a question.\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1PrecursorCorrection.html\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e793c80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HighResPrecursorMassCorrector warning: at least one feature has no convex hull - omitting feature for matching\n"
     ]
    }
   ],
   "source": [
    "PrecursorCorrection.correctToNearestFeature(feature_map_DEC, exp, 0.0, 100.0, True, False, False, False, 3, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26bd9800",
   "metadata": {},
   "source": [
    "### SIRIUS Adapter\n",
    "The SIRIUS function is optional and includes the SIRIUS Adapter Algorithm from the Boecher lab. \n",
    "\n",
    "\n",
    "The algorithm generates formula prediction from scores calculated from 1) MS2 fragmentation scores (ppm error + intensity) and 2) MS1 isotopic pattern scores.\n",
    "\n",
    "\n",
    "It can only compute feautures that are singly charged. There is also a timeout for compounds (compound timeout so that it doesn't compute for longer than 100 seconds per feature, which normally happens with larger molecules).\n",
    "\n",
    "\n",
    "-sirius:compound_timeout <number>                    \n",
    "Maximal computation time in seconds for a single compound. 0 for an infinite amount of time. (default: '100' min: '0')\n",
    "\n",
    "    \n",
    "This algorith can help in data dereplication and analysis for direct library search. \n",
    "\n",
    "#### Documentation: https://boecker-lab.github.io/docs.sirius.github.io/\n",
    "\n",
    "###### Citation: Kai Dührkop, Markus Fleischauer, Marcus Ludwig, Alexander A. Aksenov, Alexey V. Melnik, Marvin Meusel, Pieter C. Dorrestein, Juho Rousu, and Sebastian Böcker, SIRIUS 4: Turning tandem mass spectra into metabolite structure information. Nature Methods 16, 299–302, 2019."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e2bd60",
   "metadata": {},
   "source": [
    "We need to preprocess (FeatureFinderMetabo) again without convex hulls this time, otherwise the results from SIRIUS are incorrect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cc2c4e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_map_FFM = FeatureMap()\n",
    "feat_chrom = []\n",
    "ffm = FeatureFindingMetabo()\n",
    "ffm_par = ffm.getDefaults() \n",
    "ffm_par.setValue(\"isotope_filtering_model\", \"none\")\n",
    "ffm_par.setValue(\"remove_single_traces\", \"false\")\n",
    "ffm_par.setValue(\"mz_scoring_by_elements\", \"false\")\n",
    "ffm.setParameters(ffm_par)\n",
    "ffm.run(mass_traces_final, feature_map_FFM, feat_chrom)\n",
    "feature_map_FFM.setUniqueIds()\n",
    "fh = FeatureXMLFile()\n",
    "filename= \"results/Epemicins/interim/precursorcorrected_Epemicins.mzML\"\n",
    "feature_map_FFM.setPrimaryMSRunPath([filename.encode()])\n",
    "fh.store('./results/Epemicins/interim/FFM_noconvexhulls_Epemicins.featureXML', feature_map_FFM)\n",
    "mfd = MetaboliteFeatureDeconvolution()\n",
    "mdf_par = mfd.getDefaults()\n",
    "mdf_par.setValue(\"potential_adducts\",  [b\"H:+:0.6\",b\"Na:+:0.2\",b\"NH4:+:0.1\", b\"H2O:-:0.1\"])\n",
    "mdf_par.setValue(\"charge_min\", 1, \"Minimal possible charge\")\n",
    "mdf_par.setValue(\"charge_max\", 1, \"Maximal possible charge\")\n",
    "mdf_par.setValue(\"charge_span_max\", 1)\n",
    "mdf_par.setValue(\"max_neutrals\", 1)\n",
    "mfd.setParameters(mdf_par)\n",
    "feature_map_DEC = FeatureMap()\n",
    "cons_map0 = ConsensusMap()\n",
    "cons_map1 = ConsensusMap()\n",
    "mfd.compute(feature_map_FFM, feature_map_DEC, cons_map0, cons_map1)\n",
    "fxml = FeatureXMLFile()\n",
    "fh.store('./results/Epemicins/interim/MFD_noconvexhulls_Epemicins.featureXML', feature_map_DEC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38c19764",
   "metadata": {},
   "outputs": [],
   "source": [
    "sirius_algo = SiriusAdapterAlgorithm()\n",
    "sirius_algo_par = sirius_algo.getDefaults()\n",
    "sirius_algo_par.setValue(\"preprocessing:filter_by_num_masstraces\", 2) #Number of mass traces each feature has to have to be included\n",
    "sirius_algo_par.setValue(\"preprocessing:feature_only\", \"true\") #Uses the feature information from in_featureinfo to reduce the search space to MS2\n",
    "sirius_algo_par.setValue(\"sirius:profile\", \"orbitrap\")\n",
    "sirius_algo_par.setValue(\"sirius:db\", \"none\")\n",
    "sirius_algo_par.setValue(\"sirius:ions_considered\", \"[M+H]+, [M-H2O+H]+, [M+Na]+, [M+NH4]+\")\n",
    "sirius_algo_par.setValue(\"sirius:elements_enforced\", \"CHN[15]OS[4]\") \n",
    "sirius_algo_par.setValue(\"project:processors\", 2)\n",
    "sirius_algo_par.setValue(\"fingerid:db\", \"BIO\")\n",
    "sirius_algo.setParameters(sirius_algo_par)\n",
    "    \n",
    "featureinfo = \"./results/Epemicins/interim/MFD_noconvexhulls_Epemicins.featureXML\"\n",
    "fm_info = FeatureMapping_FeatureMappingInfo()\n",
    "feature_mapping = FeatureMapping_FeatureToMs2Indices() \n",
    "sirius_algo.preprocessingSirius(featureinfo,\n",
    "                                exp,\n",
    "                                fm_info,\n",
    "                                feature_mapping)\n",
    "sirius_algo.logFeatureSpectraNumber(featureinfo, \n",
    "                                    feature_mapping,\n",
    "                                    exp)\n",
    "msfile = SiriusMSFile()\n",
    "debug_level = 3\n",
    "sirius_tmp = SiriusTemporaryFileSystemObjects(debug_level)\n",
    "siriusstring= String(sirius_tmp.getTmpMsFile())\n",
    "feature_only = sirius_algo.isFeatureOnly()\n",
    "isotope_pattern_iterations = sirius_algo.getIsotopePatternIterations()\n",
    "no_mt_info = sirius_algo.isNoMasstraceInfoIsotopePattern()\n",
    "compound_info = []\n",
    "msfile.store(exp,\n",
    "             String(sirius_tmp.getTmpMsFile()),\n",
    "             feature_mapping, \n",
    "             feature_only,\n",
    "             isotope_pattern_iterations, \n",
    "             no_mt_info, \n",
    "             compound_info)\n",
    "out_csifingerid = \"./results/Epemicins/interim/structures_Epemicins.mzTab\" \n",
    "executable= \"/path/to/sirius_executable\"\n",
    "subdirs = sirius_algo.callSiriusQProcess(String(sirius_tmp.getTmpMsFile()),\n",
    "                                         String(sirius_tmp.getTmpOutDir()),\n",
    "                                         String(executable),\n",
    "                                         String(out_csifingerid),\n",
    "                                         False)\n",
    "candidates = sirius_algo.getNumberOfSiriusCandidates()\n",
    "sirius_result = MzTab()\n",
    "siriusfile = MzTabFile()\n",
    "SiriusMzTabWriter.read(subdirs,\n",
    "                       filename,\n",
    "                       candidates,\n",
    "                       sirius_result)\n",
    "siriusfile.store(\"./results/Epemicins/interim/formulas_Epemicins.mzTab\", sirius_result) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3d00a50f-7f8b-4e37-b06b-e25af34de8f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chemical_formula</th>\n",
       "      <th>exp_mass_to_charge</th>\n",
       "      <th>retention_time</th>\n",
       "      <th>best_search_engine_score[1]</th>\n",
       "      <th>best_search_engine_score[2]</th>\n",
       "      <th>best_search_engine_score[3]</th>\n",
       "      <th>opt_global_adduct</th>\n",
       "      <th>opt_gobal_precursorFormula</th>\n",
       "      <th>opt_global_rank</th>\n",
       "      <th>opt_global_explainedPeaks</th>\n",
       "      <th>opt_global_explainedIntensity</th>\n",
       "      <th>opt_global_median_mass_error_fragment_peaks_ppm</th>\n",
       "      <th>opt_global_median_absolute_mass_error_fragment_peaks_ppm</th>\n",
       "      <th>opt_global_mass_error_precursor_ppm</th>\n",
       "      <th>opt_global_compoundId</th>\n",
       "      <th>opt_global_compoundScanNumber</th>\n",
       "      <th>opt_global_featureId</th>\n",
       "      <th>opt_global_native_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2782</th>\n",
       "      <td>C7H16N3OS</td>\n",
       "      <td>208.133234</td>\n",
       "      <td>329.607460</td>\n",
       "      <td>13.604441</td>\n",
       "      <td>13.604441</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C7H20N3OS</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.935599</td>\n",
       "      <td>6.190684</td>\n",
       "      <td>8.389324</td>\n",
       "      <td>-9.606428</td>\n",
       "      <td>1576</td>\n",
       "      <td>1577</td>\n",
       "      <td>id_11804993995459637598</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1577|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2781</th>\n",
       "      <td>C10H12N3O</td>\n",
       "      <td>208.133234</td>\n",
       "      <td>329.607460</td>\n",
       "      <td>15.230887</td>\n",
       "      <td>14.193324</td>\n",
       "      <td>1.037563</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C10H16N3O</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.935599</td>\n",
       "      <td>6.881382</td>\n",
       "      <td>6.881382</td>\n",
       "      <td>6.590543</td>\n",
       "      <td>1576</td>\n",
       "      <td>1577</td>\n",
       "      <td>id_11804993995459637598</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1577|...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3687</th>\n",
       "      <td>C14H13NO3</td>\n",
       "      <td>261.123425</td>\n",
       "      <td>228.511545</td>\n",
       "      <td>46.917919</td>\n",
       "      <td>46.549750</td>\n",
       "      <td>0.368169</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C14H17NO3</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>0.833473</td>\n",
       "      <td>-1.250390</td>\n",
       "      <td>1.250390</td>\n",
       "      <td>0.215089</td>\n",
       "      <td>1057</td>\n",
       "      <td>1058</td>\n",
       "      <td>id_14669583106630530835</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3688</th>\n",
       "      <td>C12H11N4O2</td>\n",
       "      <td>261.123425</td>\n",
       "      <td>228.511545</td>\n",
       "      <td>39.257591</td>\n",
       "      <td>39.257591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C12H15N4O2</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>0.996021</td>\n",
       "      <td>5.349610</td>\n",
       "      <td>5.349610</td>\n",
       "      <td>5.357041</td>\n",
       "      <td>1057</td>\n",
       "      <td>1058</td>\n",
       "      <td>id_14669583106630530835</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3689</th>\n",
       "      <td>C7H13N7OS</td>\n",
       "      <td>261.123425</td>\n",
       "      <td>228.511545</td>\n",
       "      <td>34.713566</td>\n",
       "      <td>34.713566</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C7H17N7OS</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>0.982498</td>\n",
       "      <td>7.144945</td>\n",
       "      <td>7.144945</td>\n",
       "      <td>-2.411102</td>\n",
       "      <td>1057</td>\n",
       "      <td>1058</td>\n",
       "      <td>id_14669583106630530835</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551</th>\n",
       "      <td>C45H22N12O24S10</td>\n",
       "      <td>1451.827148</td>\n",
       "      <td>415.963700</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C45H26N12O24S10</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.756276</td>\n",
       "      <td>0.469991</td>\n",
       "      <td>2.411584</td>\n",
       "      <td>-9.902878</td>\n",
       "      <td>2076</td>\n",
       "      <td>2077</td>\n",
       "      <td>id_16989201925219475051</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>C31H26N10O40S8</td>\n",
       "      <td>1451.827148</td>\n",
       "      <td>415.963700</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C31H30N10O40S8</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>0.756276</td>\n",
       "      <td>0.469991</td>\n",
       "      <td>2.411584</td>\n",
       "      <td>-9.657702</td>\n",
       "      <td>2076</td>\n",
       "      <td>2077</td>\n",
       "      <td>id_16989201925219475051</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>C38H38N10O20S15</td>\n",
       "      <td>1451.827148</td>\n",
       "      <td>415.963700</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C38H42N10O20S15</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>0.756276</td>\n",
       "      <td>0.469991</td>\n",
       "      <td>2.411584</td>\n",
       "      <td>-9.730014</td>\n",
       "      <td>2076</td>\n",
       "      <td>2077</td>\n",
       "      <td>id_16989201925219475051</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>C30H42N12O22S16</td>\n",
       "      <td>1451.827148</td>\n",
       "      <td>415.963700</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C30H46N12O22S16</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>0.756276</td>\n",
       "      <td>0.469991</td>\n",
       "      <td>2.411584</td>\n",
       "      <td>-9.281178</td>\n",
       "      <td>2076</td>\n",
       "      <td>2077</td>\n",
       "      <td>id_16989201925219475051</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>C37H26N14O26S11</td>\n",
       "      <td>1451.827148</td>\n",
       "      <td>415.963700</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>29.538484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C37H30N14O26S11</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>0.756276</td>\n",
       "      <td>0.469991</td>\n",
       "      <td>2.411584</td>\n",
       "      <td>-9.454042</td>\n",
       "      <td>2076</td>\n",
       "      <td>2077</td>\n",
       "      <td>id_16989201925219475051</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2077</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2707 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chemical_formula  exp_mass_to_charge  retention_time  \\\n",
       "2782        C7H16N3OS          208.133234      329.607460   \n",
       "2781        C10H12N3O          208.133234      329.607460   \n",
       "3687        C14H13NO3          261.123425      228.511545   \n",
       "3688       C12H11N4O2          261.123425      228.511545   \n",
       "3689        C7H13N7OS          261.123425      228.511545   \n",
       "...               ...                 ...             ...   \n",
       "551   C45H22N12O24S10         1451.827148      415.963700   \n",
       "555    C31H26N10O40S8         1451.827148      415.963700   \n",
       "554   C38H38N10O20S15         1451.827148      415.963700   \n",
       "552   C30H42N12O22S16         1451.827148      415.963700   \n",
       "553   C37H26N14O26S11         1451.827148      415.963700   \n",
       "\n",
       "      best_search_engine_score[1]  best_search_engine_score[2]  \\\n",
       "2782                    13.604441                    13.604441   \n",
       "2781                    15.230887                    14.193324   \n",
       "3687                    46.917919                    46.549750   \n",
       "3688                    39.257591                    39.257591   \n",
       "3689                    34.713566                    34.713566   \n",
       "...                           ...                          ...   \n",
       "551                     29.538484                    29.538484   \n",
       "555                     29.538484                    29.538484   \n",
       "554                     29.538484                    29.538484   \n",
       "552                     29.538484                    29.538484   \n",
       "553                     29.538484                    29.538484   \n",
       "\n",
       "      best_search_engine_score[3] opt_global_adduct  \\\n",
       "2782                     0.000000     [M + H4 + N]+   \n",
       "2781                     1.037563     [M + H4 + N]+   \n",
       "3687                     0.368169     [M + H4 + N]+   \n",
       "3688                     0.000000     [M + H4 + N]+   \n",
       "3689                     0.000000     [M + H4 + N]+   \n",
       "...                           ...               ...   \n",
       "551                      0.000000     [M + H4 + N]+   \n",
       "555                      0.000000     [M + H4 + N]+   \n",
       "554                      0.000000     [M + H4 + N]+   \n",
       "552                      0.000000     [M + H4 + N]+   \n",
       "553                      0.000000     [M + H4 + N]+   \n",
       "\n",
       "     opt_gobal_precursorFormula  opt_global_rank  opt_global_explainedPeaks  \\\n",
       "2782                  C7H20N3OS                3                          4   \n",
       "2781                  C10H16N3O                2                          4   \n",
       "3687                  C14H17NO3                1                         11   \n",
       "3688                 C12H15N4O2                2                         14   \n",
       "3689                  C7H17N7OS                3                         13   \n",
       "...                         ...              ...                        ...   \n",
       "551             C45H26N12O24S10                1                         12   \n",
       "555              C31H30N10O40S8                5                         12   \n",
       "554             C38H42N10O20S15                4                         12   \n",
       "552             C30H46N12O22S16                2                         12   \n",
       "553             C37H30N14O26S11                3                         12   \n",
       "\n",
       "      opt_global_explainedIntensity  \\\n",
       "2782                       0.935599   \n",
       "2781                       0.935599   \n",
       "3687                       0.833473   \n",
       "3688                       0.996021   \n",
       "3689                       0.982498   \n",
       "...                             ...   \n",
       "551                        0.756276   \n",
       "555                        0.756276   \n",
       "554                        0.756276   \n",
       "552                        0.756276   \n",
       "553                        0.756276   \n",
       "\n",
       "      opt_global_median_mass_error_fragment_peaks_ppm  \\\n",
       "2782                                         6.190684   \n",
       "2781                                         6.881382   \n",
       "3687                                        -1.250390   \n",
       "3688                                         5.349610   \n",
       "3689                                         7.144945   \n",
       "...                                               ...   \n",
       "551                                          0.469991   \n",
       "555                                          0.469991   \n",
       "554                                          0.469991   \n",
       "552                                          0.469991   \n",
       "553                                          0.469991   \n",
       "\n",
       "      opt_global_median_absolute_mass_error_fragment_peaks_ppm  \\\n",
       "2782                                           8.389324          \n",
       "2781                                           6.881382          \n",
       "3687                                           1.250390          \n",
       "3688                                           5.349610          \n",
       "3689                                           7.144945          \n",
       "...                                                 ...          \n",
       "551                                            2.411584          \n",
       "555                                            2.411584          \n",
       "554                                            2.411584          \n",
       "552                                            2.411584          \n",
       "553                                            2.411584          \n",
       "\n",
       "      opt_global_mass_error_precursor_ppm  opt_global_compoundId  \\\n",
       "2782                            -9.606428                   1576   \n",
       "2781                             6.590543                   1576   \n",
       "3687                             0.215089                   1057   \n",
       "3688                             5.357041                   1057   \n",
       "3689                            -2.411102                   1057   \n",
       "...                                   ...                    ...   \n",
       "551                             -9.902878                   2076   \n",
       "555                             -9.657702                   2076   \n",
       "554                             -9.730014                   2076   \n",
       "552                             -9.281178                   2076   \n",
       "553                             -9.454042                   2076   \n",
       "\n",
       "      opt_global_compoundScanNumber     opt_global_featureId  \\\n",
       "2782                           1577  id_11804993995459637598   \n",
       "2781                           1577  id_11804993995459637598   \n",
       "3687                           1058  id_14669583106630530835   \n",
       "3688                           1058  id_14669583106630530835   \n",
       "3689                           1058  id_14669583106630530835   \n",
       "...                             ...                      ...   \n",
       "551                            2077  id_16989201925219475051   \n",
       "555                            2077  id_16989201925219475051   \n",
       "554                            2077  id_16989201925219475051   \n",
       "552                            2077  id_16989201925219475051   \n",
       "553                            2077  id_16989201925219475051   \n",
       "\n",
       "                                   opt_global_native_id  \n",
       "2782  controllerType=0 controllerNumber=1 scan=1577|...  \n",
       "2781  controllerType=0 controllerNumber=1 scan=1577|...  \n",
       "3687      controllerType=0 controllerNumber=1 scan=1058  \n",
       "3688      controllerType=0 controllerNumber=1 scan=1058  \n",
       "3689      controllerType=0 controllerNumber=1 scan=1058  \n",
       "...                                                 ...  \n",
       "551       controllerType=0 controllerNumber=1 scan=2077  \n",
       "555       controllerType=0 controllerNumber=1 scan=2077  \n",
       "554       controllerType=0 controllerNumber=1 scan=2077  \n",
       "552       controllerType=0 controllerNumber=1 scan=2077  \n",
       "553       controllerType=0 controllerNumber=1 scan=2077  \n",
       "\n",
       "[2707 rows x 18 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "siriusfile= \"./results/Epemicins/interim/formulas_Epemicins.mzTab\"\n",
    "sirius=  pyteomics.mztab.MzTab(siriusfile, encoding='UTF8', table_format='df')\n",
    "sirius.metadata\n",
    "df= sirius.small_molecule_table\n",
    "SIRIUS_DF= df.drop(columns= [\"identifier\", \"smiles\", \"inchi_key\", \"description\", \"calc_mass_to_charge\", \"charge\", \"taxid\", \"species\",\"database\", \"database_version\", \"spectra_ref\", \"search_engine\", \"modifications\"])\n",
    "SIRIUS_DF.to_csv(\"./results/Epemicins/formulas_Epemicins.csv\")\n",
    "SIRIUS_DF= pd.read_csv(\"./results/Epemicins/formulas_Epemicins.csv\", index_col= \"Unnamed: 0\")\n",
    "SIRIUS_DF=SIRIUS_DF[SIRIUS_DF[\"opt_global_explainedIntensity\"] >= 0.6] #opt_global_explainedIntensity should be higher than 0.8 or 0.9 even for reliable results\n",
    "SIRIUS_DF= SIRIUS_DF.sort_values(by= \"exp_mass_to_charge\")\n",
    "SIRIUS_DF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d4c0d4e-ffc1-4865-a9e5-129f3d3c0592",
   "metadata": {},
   "source": [
    "### CSIFinger:ID\n",
    "\n",
    "The CSI_fingerID function is another algorithm from the Boecher lab, just like SIRIUS adapter and is using the formula predictions from SIRIUS, to search in structural libraries and predict the structure of each formula\n",
    "\n",
    "#### Documentation: https://boecker-lab.github.io/docs.sirius.github.io/\n",
    "\n",
    "##### Citation: Kai Dührkop, Huibin Shen, Marvin Meusel, Juho Rousu, and Sebastian Böcker, Searching molecular structure databases with tandem mass spectra using CSI:FingerID, PNAS October 13, 2015 112 (41) 12580-12585, https://doi.org/10.1073/pnas.1509788112"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "261217e8-39a3-4ae4-9e22-18b110645b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "csi_file.store(\"./results/Epemicins/interim/structures_Epemicins.mzTab\", csi_result)\n",
    "csi_file= \"./results/Epemicins/interim/structures_Epemicins.mzTab\"\n",
    "CSI=  pyteomics.mztab.MzTab(csi_file, encoding='UTF8', table_format='df')\n",
    "CSI.metadata\n",
    "df= CSI.small_molecule_table\n",
    "csifingerID= df.drop(columns= [\"calc_mass_to_charge\", \"charge\", \"taxid\", \"species\",\"database\", \"database_version\", \"spectra_ref\", \"search_engine\", \"modifications\"])\n",
    "csifingerID.to_csv(\"./results/Epemicins/structures_Epemicins.csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04f80b6f",
   "metadata": {},
   "source": [
    "### Explanation of columns\n",
    "##### mz= mass-to-charge ratio (m/z)\n",
    "##### RT= retention time (sec)\n",
    "##### intensity = intensity of the feature (AU-arbitrary units)\n",
    "##### FWHM= Full Width of the peak at Half its Maximum height\n",
    "##### num_of_masstraces\t= number of mass traces detected (single mass traces are excluded). This is relevant to the isotopic pattern\n",
    "##### isotope_distances = distance in mz between the isotopes (jumps of app. 1 is important to confirm that this is a real feature) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "db72257d-802c-4fcf-99f0-19e81e7398e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>chemical_formula</th>\n",
       "      <th>smiles</th>\n",
       "      <th>inchi_key</th>\n",
       "      <th>description</th>\n",
       "      <th>exp_mass_to_charge</th>\n",
       "      <th>retention_time</th>\n",
       "      <th>best_search_engine_score[1]</th>\n",
       "      <th>opt_global_rank</th>\n",
       "      <th>opt_global_compoundId</th>\n",
       "      <th>opt_global_compoundScanNumber</th>\n",
       "      <th>opt_global_featureId</th>\n",
       "      <th>opt_global_native_id</th>\n",
       "      <th>opt_global_adduct</th>\n",
       "      <th>opt_global_dblinks</th>\n",
       "      <th>opt_global_dbflags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14852549|14852550|25053655|24971283|53652498|5...</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCCCCCCC(CC1C(C(=O)O1)CCCCCC)OC(=O)CNC=O</td>\n",
       "      <td>ANPULBVRORHAGO</td>\n",
       "      <td>Orlistat derivative, 15g</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-167.839042</td>\n",
       "      <td>1</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>MeSH:(14852550)|PubChem:(14852549 14852550 250...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9889418|11059239</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCCCCCC1C(OC1=O)CC(CCCCCCC)OC(=O)CNC=O</td>\n",
       "      <td>RQFZMIXBLJEUGS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-168.290573</td>\n",
       "      <td>2</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>COCONUT:(CNP0260950)|KNApSAcK:(16423)|Natural ...</td>\n",
       "      <td>3178514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>71464556|71464557</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCC=CCC=CCCCCCC(CC(=O)OC(CC(=O)O)C[N+](C)(C...</td>\n",
       "      <td>WQYXCASYXUFNSI</td>\n",
       "      <td>3-hydroxylinoleylcarnitine</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-170.979093</td>\n",
       "      <td>3</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>PubChem:(71464556 71464557)|CHEBI:(73075)</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9824494</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCCC(CC1C(C(=O)O1)CCCCCCCC(C)C)OC(=O)CNC=O</td>\n",
       "      <td>SMEWLQMERHEMJK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-171.211550</td>\n",
       "      <td>4</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>COCONUT:(CNP0157254)|Natural Products:(UNPD174...</td>\n",
       "      <td>3178498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3049003</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCC(C)(CCN1C(C(=O)C(C1=O)(C)C)CCCCCCC(=O)O...</td>\n",
       "      <td>OHQKFFPFFFUNJW</td>\n",
       "      <td>Ethyl 4,4-dimethyl-3,5-dioxo-1-(3-hydroxy-3-me...</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-250.749662</td>\n",
       "      <td>5</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>PubChem:(3049003)|PubChem class - safety and t...</td>\n",
       "      <td>67108866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2572</th>\n",
       "      <td>288197|2754113|7056395|7056399|11868690|118686...</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC12CCC3C(C1CCC2=O)CCC(C3(C)CC(=O)O)CC(=O)O</td>\n",
       "      <td>CEUQHWSIXGFNGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-155.792453</td>\n",
       "      <td>1</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0068730)|PubChem:(288197 2754113 7...</td>\n",
       "      <td>3153922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>\"\"</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC(=O)C(=O)C1(CCC2C(C(=O)CCC2(C1CC(=O)O)C)(C)C)C</td>\n",
       "      <td>QWLGLRAGSFVBDI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-168.039526</td>\n",
       "      <td>2</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0136763)|Natural Products:(UNPD195...</td>\n",
       "      <td>3178496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2574</th>\n",
       "      <td>354385|494326|23425264</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC(=O)OC1CC2CC3(C(CC(C3C2(C1=C)O)(C)C)OC(=O)C)C</td>\n",
       "      <td>CLIPYJQRTQNGAM</td>\n",
       "      <td>.delta.-(sup 9,8.beta., 10.alpha.-triol, 3,8-d...</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-203.768295</td>\n",
       "      <td>3</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0216092)|Natural Products:(UNPD156...</td>\n",
       "      <td>3178498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2575</th>\n",
       "      <td>\"\"</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC(=O)OC1CC2CC3(CC(C(C3C2(C1=C)O)(C)C)OC(=O)C)C</td>\n",
       "      <td>KEVCQSXXJWIVSR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-205.873518</td>\n",
       "      <td>4</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0261220)|Natural Products:(UNPD183...</td>\n",
       "      <td>2129920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576</th>\n",
       "      <td>11717027</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CCCCCCCC(C(C(C#CC#CC(C=C)O)O)OC(=O)C)O</td>\n",
       "      <td>TYSPCPWTSAFCRA</td>\n",
       "      <td>[(8R,9R,10R)-3,8,10-trihydroxyheptadec-1-en-4,...</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-210.247162</td>\n",
       "      <td>5</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0186238)|Natural Products:(UNPD125...</td>\n",
       "      <td>2129922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2577 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             identifier chemical_formula  \\\n",
       "0     14852549|14852550|25053655|24971283|53652498|5...        C25H45NO5   \n",
       "1                                      9889418|11059239        C25H45NO5   \n",
       "2                                     71464556|71464557        C25H45NO5   \n",
       "3                                               9824494        C25H45NO5   \n",
       "4                                               3049003        C25H45NO5   \n",
       "...                                                 ...              ...   \n",
       "2572  288197|2754113|7056395|7056399|11868690|118686...         C19H28O5   \n",
       "2573                                                 \"\"         C19H28O5   \n",
       "2574                             354385|494326|23425264         C19H28O5   \n",
       "2575                                                 \"\"         C19H28O5   \n",
       "2576                                           11717027         C19H28O5   \n",
       "\n",
       "                                                 smiles       inchi_key  \\\n",
       "0          CCCCCCCCCCCC(CC1C(C(=O)O1)CCCCCC)OC(=O)CNC=O  ANPULBVRORHAGO   \n",
       "1            CCCCCCCCCCC1C(OC1=O)CC(CCCCCCC)OC(=O)CNC=O  RQFZMIXBLJEUGS   \n",
       "2     CCCCCC=CCC=CCCCCCC(CC(=O)OC(CC(=O)O)C[N+](C)(C...  WQYXCASYXUFNSI   \n",
       "3        CCCCCCCC(CC1C(C(=O)O1)CCCCCCCC(C)C)OC(=O)CNC=O  SMEWLQMERHEMJK   \n",
       "4     CCCCCCC(C)(CCN1C(C(=O)C(C1=O)(C)C)CCCCCCC(=O)O...  OHQKFFPFFFUNJW   \n",
       "...                                                 ...             ...   \n",
       "2572        CC12CCC3C(C1CCC2=O)CCC(C3(C)CC(=O)O)CC(=O)O  CEUQHWSIXGFNGI   \n",
       "2573   CC(=O)C(=O)C1(CCC2C(C(=O)CCC2(C1CC(=O)O)C)(C)C)C  QWLGLRAGSFVBDI   \n",
       "2574    CC(=O)OC1CC2CC3(C(CC(C3C2(C1=C)O)(C)C)OC(=O)C)C  CLIPYJQRTQNGAM   \n",
       "2575    CC(=O)OC1CC2CC3(CC(C(C3C2(C1=C)O)(C)C)OC(=O)C)C  KEVCQSXXJWIVSR   \n",
       "2576             CCCCCCCC(C(C(C#CC#CC(C=C)O)O)OC(=O)C)O  TYSPCPWTSAFCRA   \n",
       "\n",
       "                                            description  exp_mass_to_charge  \\\n",
       "0                              Orlistat derivative, 15g          457.363953   \n",
       "1                                                   NaN          457.363953   \n",
       "2                            3-hydroxylinoleylcarnitine          457.363953   \n",
       "3                                                   NaN          457.363953   \n",
       "4     Ethyl 4,4-dimethyl-3,5-dioxo-1-(3-hydroxy-3-me...          457.363953   \n",
       "...                                                 ...                 ...   \n",
       "2572                                                NaN          337.199054   \n",
       "2573                                                NaN          337.199054   \n",
       "2574  .delta.-(sup 9,8.beta., 10.alpha.-triol, 3,8-d...          337.199054   \n",
       "2575                                                NaN          337.199054   \n",
       "2576  [(8R,9R,10R)-3,8,10-trihydroxyheptadec-1-en-4,...          337.199054   \n",
       "\n",
       "      retention_time  best_search_engine_score[1]  opt_global_rank  \\\n",
       "0         327.701974                  -167.839042                1   \n",
       "1         327.701974                  -168.290573                2   \n",
       "2         327.701974                  -170.979093                3   \n",
       "3         327.701974                  -171.211550                4   \n",
       "4         327.701974                  -250.749662                5   \n",
       "...              ...                          ...              ...   \n",
       "2572      465.541832                  -155.792453                1   \n",
       "2573      465.541832                  -168.039526                2   \n",
       "2574      465.541832                  -203.768295                3   \n",
       "2575      465.541832                  -205.873518                4   \n",
       "2576      465.541832                  -210.247162                5   \n",
       "\n",
       "      opt_global_compoundId  opt_global_compoundScanNumber  \\\n",
       "0                      1571                           1572   \n",
       "1                      1571                           1572   \n",
       "2                      1571                           1572   \n",
       "3                      1571                           1572   \n",
       "4                      1571                           1572   \n",
       "...                     ...                            ...   \n",
       "2572                   2336                           2337   \n",
       "2573                   2336                           2337   \n",
       "2574                   2336                           2337   \n",
       "2575                   2336                           2337   \n",
       "2576                   2336                           2337   \n",
       "\n",
       "        opt_global_featureId                           opt_global_native_id  \\\n",
       "0     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "1     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "2     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "3     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "4     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "...                      ...                                            ...   \n",
       "2572   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2573   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2574   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2575   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2576   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "\n",
       "     opt_global_adduct                                 opt_global_dblinks  \\\n",
       "0        [M + H4 + N]+  MeSH:(14852550)|PubChem:(14852549 14852550 250...   \n",
       "1        [M + H4 + N]+  COCONUT:(CNP0260950)|KNApSAcK:(16423)|Natural ...   \n",
       "2        [M + H4 + N]+          PubChem:(71464556 71464557)|CHEBI:(73075)   \n",
       "3        [M + H4 + N]+  COCONUT:(CNP0157254)|Natural Products:(UNPD174...   \n",
       "4        [M + H4 + N]+  PubChem:(3049003)|PubChem class - safety and t...   \n",
       "...                ...                                                ...   \n",
       "2572          [M + H]+  COCONUT:(CNP0068730)|PubChem:(288197 2754113 7...   \n",
       "2573          [M + H]+  COCONUT:(CNP0136763)|Natural Products:(UNPD195...   \n",
       "2574          [M + H]+  COCONUT:(CNP0216092)|Natural Products:(UNPD156...   \n",
       "2575          [M + H]+  COCONUT:(CNP0261220)|Natural Products:(UNPD183...   \n",
       "2576          [M + H]+  COCONUT:(CNP0186238)|Natural Products:(UNPD125...   \n",
       "\n",
       "      opt_global_dbflags  \n",
       "0                     70  \n",
       "1                3178514  \n",
       "2                     34  \n",
       "3                3178498  \n",
       "4               67108866  \n",
       "...                  ...  \n",
       "2572             3153922  \n",
       "2573             3178496  \n",
       "2574             3178498  \n",
       "2575             2129920  \n",
       "2576             2129922  \n",
       "\n",
       "[2577 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example\n",
    "csifingerID= pd.read_csv(\"./results/Epemicins/structures_Epemicins.csv\",index_col= \"Unnamed: 0\")\n",
    "csifingerID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d0b7749-07f7-48fa-85c7-afa5561709b6",
   "metadata": {},
   "source": [
    "### Create a metadata csv file for GNPS from the samples.tsv file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4d147d81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>ATTRIBUTE_comment</th>\n",
       "      <th>ATTRIBUTE_genomeID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Epemicins.mzml</td>\n",
       "      <td>MAP0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FilipinIII.mzml</td>\n",
       "      <td>MAP1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GermicidinA.mzml</td>\n",
       "      <td>MAP2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GermicidinB.mzml</td>\n",
       "      <td>MAP3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Leupeptin.mzml</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pentamycin.mzml</td>\n",
       "      <td>MAP5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           filename ATTRIBUTE_comment  ATTRIBUTE_genomeID\n",
       "0    Epemicins.mzml              MAP0                 NaN\n",
       "1   FilipinIII.mzml              MAP1                 NaN\n",
       "2  GermicidinA.mzml              MAP2                 NaN\n",
       "3  GermicidinB.mzml              MAP3                 NaN\n",
       "4    Leupeptin.mzml              MAP4                 NaN\n",
       "5   Pentamycin.mzml              MAP5                 NaN"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path= \"results/GNPSexport/\"\n",
    "isExist= os.path.exists(path)\n",
    "if not isExist:\n",
    "    os.mkdir(\"results/GNPSexport/\")\n",
    "df= pd.read_csv(\"config/samples.tsv\", sep= \"\\t\", index_col= \"Unnamed: 0\")\n",
    "metadata= df.rename(columns= {\"sample_name\": \"filename\", \"comment\": \"ATTRIBUTE_comment\", \"MAPnumber\": \"ATTRIBUTE_MAPnumber\"})\n",
    "metadata[\"filename\"]= metadata[\"filename\"].astype(str) +\".mzml\"\n",
    "metadata['ATTRIBUTE_MAPnumber'] = np.arange(len(metadata))\n",
    "metadata[\"ATTRIBUTE_comment\"]= \"MAP\" + metadata[\"ATTRIBUTE_MAPnumber\"].astype(str)\n",
    "metadata= metadata.drop(columns= \"ATTRIBUTE_MAPnumber\")\n",
    "metadata['ATTRIBUTE_genomeID']=metadata['filename'].str.extract(r'(NBC_?\\d*)')\n",
    "metadata['ATTRIBUTE_genomeIDMDNA']=metadata['filename'].str.extract(r'(MDNA_WGS_?\\d*)')\n",
    "metadata['ATTRIBUTE_genomeID']=metadata['ATTRIBUTE_genomeID'].fillna(metadata['ATTRIBUTE_genomeIDMDNA'])\n",
    "metadata=metadata.drop(columns=\"ATTRIBUTE_genomeIDMDNA\")\n",
    "#metadata['ATTRIBUTE_genomeID']= metadata['ATTRIBUTE_genomeID'].replace(to_replace= r'NBC', value= 'NBC_', regex= True)\n",
    "#metadata['ATTRIBUTE_genomeID']= metadata['ATTRIBUTE_genomeID'].replace(to_replace= r'MDNAWGS', value= 'MDNA_WGS_', regex= True)\n",
    "metadata.to_csv(\"results/GNPSexport/metadata.tsv\", sep='\\t')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60121532-63ae-41f6-a028-d130f3aa2d80",
   "metadata": {},
   "source": [
    "From the CLI, I am using the following algorithms to generate files compatible to GNPS for FBMN: \n",
    "\n",
    "#### MapAlignerPoseClustering is used to perform a linear retention time alignment, basically correct for linear shifts in retention time. \n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_MapAlignerPoseClustering.html\n",
    "\n",
    "    rule MapAlignerPoseClustering:\n",
    "        input:\n",
    "            expand(\"results/{samples}/interim/MFD_{samples}.featureXML\", samples=SAMPLES)\n",
    "        output:\n",
    "            var1= expand(\"results/Consensus/interim/MapAlignerPoseClustering_{samples}.featureXML\", samples=SAMPLES),\n",
    "            var2= expand(\"results/Consensus/interim/MapAlignerPoseClustering_{samples}.trafoXML\", samples=SAMPLES)\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/MapAlignerPoseClustering -in {input} -out {output.var1} -trafo_out {output.var2}\n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "fe932f36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Picking a reference (by size) ... done\n",
      "Progress of 'Aligning input maps':\n",
      "-- done [took 4.47 s (CPU), 4.54 s (Wall)] -- \n",
      "MapAlignerPoseClustering took 6.10 s (wall), 6.02 s (CPU), 0.30 s (system), 5.72 s (user); Peak Memory Usage: 68 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/MapAlignerPoseClustering -algorithm:max_num_peaks_considered -1 -in results/Epemicins/interim/MFD_Epemicins.featureXML results/GermicidinA/interim/MFD_GermicidinA.featureXML results/GermicidinB/interim/MFD_GermicidinB.featureXML -out results/Consensus/interim/MapAlignerPoseClustering_Epemicins.featureXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinA.featureXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinB.featureXML -trafo_out results/Consensus/interim/MapAlignerPoseClustering_Epemicins.trafoXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinA.trafoXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinB.trafoXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e00257a6",
   "metadata": {},
   "source": [
    "#### Introduce the features to a protein identification file (idXML)   \n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_IDMapper.html\n",
    "\n",
    "    rule IDMapper:\n",
    "        input:\n",
    "            \"resources/emptyfile.idXML\",\n",
    "            \"results/Consensus/interim/MapAlignerPoseClustering_{samples}.featureXML\",\n",
    "            \"results/{samples}/interim/precursorcorrected_{samples}.mzML\"\n",
    "        output:\n",
    "            \"results/Consensus/interim/IDMapper_{samples}.featureXML\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/IDMapper -id {input[0]} -in {input[1]}  -spectra:in {input[2]} -out {output} \n",
    "            \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e0cbea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!resources/OpenMS-2.7.0/bin/IDMapper -id resources/emptyfile.idXML -in results/GNPSexport/interim/MapAlignerPoseClustering_Epemicins.featureXML  -spectra:in results/Epemicins/interim/precursorcorrected_Epemicins.mzML -out results/GNPSexport/interim/IDMapper_Epemicins.featureXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12202f9b",
   "metadata": {},
   "source": [
    "#### The FeatureLinkerUnlabeledKD is used to aggregate the feature information (from single files) into a ConsensusFeature, linking features from different files together, which have a smiliar m/z and rt (no MS2 data).\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FeatureLinkerUnlabeledKD.html\n",
    "\n",
    "    rule FeatureLinkerUnlabeledKD:\n",
    "        input:\n",
    "            expand(\"results/Consensus/interim/IDMapper_{samples}.featureXML\", samples=SAMPLES)\n",
    "        output:\n",
    "            \"results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/FeatureLinkerUnlabeledKD -in {input} -out {output} \n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec78e348",
   "metadata": {},
   "outputs": [],
   "source": [
    "!resources/OpenMS-2.7.0/bin/FeatureLinkerUnlabeledKD -in results/Consensus/interim/IDMapper_Epemicins.featureXML results/Consensus/interim/IDMapper_GermicidinA.featureXML results/Consensus/interim/IDMapper_GermicidinB.featureXML -out results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670a0817",
   "metadata": {},
   "source": [
    "#### Filter out the features that do not have an MS2 pattern\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FileFilter.html\n",
    "\n",
    "    rule FileFilter:\n",
    "        input:\n",
    "            \"results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML\"\n",
    "        output:\n",
    "            \"results/Consensus/filtered.consensusXML\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/FileFilter -id:remove_unannotated_features -in {input} -out {output} \n",
    "            \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f85a472d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!resources/OpenMS-2.7.0/bin/FileFilter -id:remove_unannotated_features -in results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML -out results/Consensus/filtered.consensusXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04145cbe",
   "metadata": {},
   "source": [
    "\n",
    "#### GNPS_export creates an mgf file with only the MS2 information of all files (introduce mzml files with spaces between them)\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_GNPSExport.html\n",
    "\n",
    "    rule GNPS_export:\n",
    "        input:\n",
    "            \"results/Consensus/filtered.consensusXML\",\n",
    "            expand(\"results/{samples}/interim/precursorcorrected_{samples}.mzML\", samples=SAMPLES)\n",
    "        output:\n",
    "            \"results/GNPSexport/MSMS.mgf\" \n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/GNPSExport -in_cm {input[0]} -in_mzml {input[1]} -out {output} \n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d29ee426",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'parsing features and ms2 identifications...':\n",
      "  1.63 %               GNPSExport took 2.62 s (wall), 2.58 s (CPU), 0.16 s (system), 2.42 s (user); Peak Memory Usage: 87 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/GNPSExport -ini resources/GNPSExport.ini -in_cm results/Consensus/filtered.consensusXML -in_mzml results/Epemicins/interim/precursorcorrected_Epemicins.mzML results/GermicidinA/interim/precursorcorrected_GermicidinA.mzML results/GermicidinB/interim/precursorcorrected_GermicidinB.mzML -out results/GNPSexport/MSMS.mgf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ac98624",
   "metadata": {},
   "source": [
    "#### export the consensusXML file to a txt file for GNPS\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_TextExporter.html\n",
    "    \n",
    "    rule txt_export:\n",
    "        input:\n",
    "            \"results/Consensus/filtered.consensusXML\"\n",
    "        output:\n",
    "            \"results/GNPSexport/FeatureQuantificationTable.txt\" \n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/TextExporter -in {input} -out {output}\n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bac296b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TextExporter took 0.09 s (wall), 0.09 s (CPU), 0.01 s (system), 0.08 s (user); Peak Memory Usage: 19 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/TextExporter -in results/Consensus/filtered.consensusXML -out results/GNPSexport/FeatureQuantificationTable.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d536ae97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RT window size calculated as 240 seconds.\n",
      "Loading targets and creating assay library...\n",
      "Loading input LC-MS data...\n",
      "Progress of 'loading spectra list':\n",
      "-- done [took 0.48 s (CPU), 0.47 s (Wall)] -- \n",
      "Progress of 'loading chromatogram list':\n",
      "-- done [took 0.00 s (CPU), 0.00 s (Wall)] -- \n",
      "Error: No sum formula given for target 'Unknown'; cannot calculate isotope distribution - using estimation method for peptides.\n",
      "<RT window size calculated as 240 seconds.> occurred 4 times\n",
      "Extracting chromatograms...\n",
      "Detecting chromatographic peaks...\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 561.357276466771054.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 562.360631304571029.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 183.177276466770991.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 184.180631304570994.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 197.117276466771017.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 198.12063130457102.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 1255.637276466771027.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 1256.640631304571116.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 139.087276466771016.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 140.090631304571019.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 155.077276466770996.\n",
      "Found no signal. The Gaussian width is probably smaller than the spacing in your chromatogram data. Try to use a bigger width. The error occurred in the chromatogram with m/z time 156.080631304571.\n",
      "Found 22 feature candidates in total.\n",
      "11 features left after selection of best candidates.\n",
      "No overlaps between features found.\n",
      "Model fitting: 4 successes, 7 failures\n",
      "Imputing model failures with a linear model based on log(rawIntensities). Slope: 0.98208, Intercept: -0.138156\n",
      "Writing final results...\n",
      "\n",
      "Summary statistics:\n",
      "17 targets specified\n",
      "11 features found\n",
      "0 features with multiple target annotations\n",
      "6 targets without features:\n",
      "- Geosmin (m=182.169998, z=1, rt=383.359985)\n",
      "- pyracrimycinA (m=138.080002, z=1, rt=50.82)\n",
      "- pyracrimycinB (m=154.070007, z=1, rt=69.540001)\n",
      "- actinomycinD (m=1254.630005, z=1, rt=75.089996)\n",
      "- GermicidinA (m=196.110001, z=1, rt=52.169998)\n",
      "- ... (1 more)\n",
      "\n",
      "FeatureFinderMetaboIdent took 0.76 s (wall), 0.74 s (CPU), 0.07 s (system), 0.67 s (user); Peak Memory Usage: 78 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/FeatureFinderMetaboIdent -in results/Epemicins/interim/precursorcorrected_Epemicins.mzML -id MetaboliteIdentification.tsv -out results/Epemicins/interim/FFMI_Epemicins.featureXML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from functools import reduce\n",
    "from pathlib import Path\n",
    "from time import perf_counter\n",
    "import sys\n",
    "\n",
    "from IPython.core.display import display\n",
    "from pandas import CategoricalDtype\n",
    "import numpy as np\n",
    "from pyopenms import *\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "class FeatureMapDF(FeatureMap):      \n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "\n",
    "    def get_fmi_df(self):\n",
    "        def gen(fmap: FeatureMap, fun):\n",
    "            for f in fmap:\n",
    "                yield from fun(f)\n",
    "\n",
    "        def extractMetaData(f: Feature):\n",
    "            metavals = []\n",
    "            types = []\n",
    "            for k in metavals:\n",
    "                if k == b\"MZ\":\n",
    "                    types.append('f')\n",
    "                elif k== b\"PeptideRef\":\n",
    "                    types.append('U100')\n",
    "                elif k== b\"masserror_ppm\":\n",
    "                    types.append('i4')\n",
    "                elif k== b\"masstrace_intensity\":\n",
    "                    types.append(\"f\")\n",
    "            # subfeatures = f.getFeatureList()  # type: list[FeatureHandle]\n",
    "            pep = f.getPeptideIdentifications()  # type: list[PeptideIdentification]\n",
    "            bb = f.getConvexHull().getBoundingBox2D()\n",
    "            for u in metavals:\n",
    "                f.getMetaValue(u)\n",
    "            if len(pep) != 0:\n",
    "                hits = pep[0].getHits()\n",
    "                if len(hits) != 0:\n",
    "                    besthit = hits[0]  # type: PeptideHit\n",
    "                    yield f.getUniqueId(), besthit.getSequence().toString(), f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"PeptideRef\"), f.getOverallQuality(), f.getIntensity()\n",
    "                else:\n",
    "                    yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"PeptideRef\"), f.getOverallQuality(), f.getIntensity()\n",
    "            else:\n",
    "                yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"PeptideRef\"), f.getOverallQuality(), f.getIntensity()\n",
    "\n",
    "        cnt = self.size()\n",
    "\n",
    "        mddtypes = [('id', np.dtype('uint64')), ('sequence', 'U200'), ('charge', 'i4'), ('RT', 'f'), ('mz', 'f'),\n",
    "                    ('RTstart', 'f'), ('RTend', 'f'), (\"PeptideRef\", 'U100'),\n",
    "                    ('quality', 'f'), ('intensity', 'f')]\n",
    "        mdarr = np.fromiter(iter=gen(self, extractMetaData), dtype=mddtypes, count=cnt)\n",
    "        df= pd.DataFrame(mdarr).set_index('id').sort_values(\"mz\").drop(columns= \"sequence\")\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>charge</th>\n",
       "      <th>RT</th>\n",
       "      <th>mz</th>\n",
       "      <th>RTstart</th>\n",
       "      <th>RTend</th>\n",
       "      <th>PeptideRef</th>\n",
       "      <th>quality</th>\n",
       "      <th>intensity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6540659878518515890</th>\n",
       "      <td>1</td>\n",
       "      <td>46.026489</td>\n",
       "      <td>402.207275</td>\n",
       "      <td>37.475689</td>\n",
       "      <td>54.632912</td>\n",
       "      <td>bioxalomycina2_m401.200012_z1_rt48.18</td>\n",
       "      <td>-2.825665</td>\n",
       "      <td>2.926461e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280489307960157370</th>\n",
       "      <td>1</td>\n",
       "      <td>103.041199</td>\n",
       "      <td>474.207275</td>\n",
       "      <td>91.933479</td>\n",
       "      <td>113.822960</td>\n",
       "      <td>Unknown_m473.200012_z1_rt44.200001</td>\n",
       "      <td>-0.763105</td>\n",
       "      <td>2.209964e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9143164355464892809</th>\n",
       "      <td>1</td>\n",
       "      <td>265.116608</td>\n",
       "      <td>566.277283</td>\n",
       "      <td>243.804138</td>\n",
       "      <td>286.463776</td>\n",
       "      <td>Coelichelin_m565.27002_z1_rt268.049988</td>\n",
       "      <td>-7.000911</td>\n",
       "      <td>1.934900e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10815806269762203077</th>\n",
       "      <td>1</td>\n",
       "      <td>373.222076</td>\n",
       "      <td>585.357300</td>\n",
       "      <td>353.652527</td>\n",
       "      <td>393.302399</td>\n",
       "      <td>dehydroxynocardamine_m584.349976_z1_rt351.070007</td>\n",
       "      <td>-7.667008</td>\n",
       "      <td>6.850766e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14786621585738311964</th>\n",
       "      <td>1</td>\n",
       "      <td>286.942780</td>\n",
       "      <td>587.337280</td>\n",
       "      <td>275.987885</td>\n",
       "      <td>310.614807</td>\n",
       "      <td>desmethylenylnocardamine_m586.330017_z1_rt339....</td>\n",
       "      <td>-7.992048</td>\n",
       "      <td>9.937172e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10021272275670129286</th>\n",
       "      <td>1</td>\n",
       "      <td>298.578583</td>\n",
       "      <td>601.357300</td>\n",
       "      <td>259.177887</td>\n",
       "      <td>319.175507</td>\n",
       "      <td>desferrioxamineE_m600.349976_z1_rt329.5</td>\n",
       "      <td>-10.806923</td>\n",
       "      <td>1.364715e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9181399712805582150</th>\n",
       "      <td>2</td>\n",
       "      <td>379.947540</td>\n",
       "      <td>712.417297</td>\n",
       "      <td>333.265137</td>\n",
       "      <td>412.114746</td>\n",
       "      <td>epemicinB_m1422.819946_z2_rt405.769989</td>\n",
       "      <td>4.456927</td>\n",
       "      <td>9.244367e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11025134690498236929</th>\n",
       "      <td>2</td>\n",
       "      <td>367.807312</td>\n",
       "      <td>793.447266</td>\n",
       "      <td>326.792847</td>\n",
       "      <td>404.532959</td>\n",
       "      <td>epemicinA_m1584.880005_z2_rt386.850006</td>\n",
       "      <td>4.391304</td>\n",
       "      <td>3.584970e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12247998505257479685</th>\n",
       "      <td>1</td>\n",
       "      <td>563.042603</td>\n",
       "      <td>797.417297</td>\n",
       "      <td>537.345642</td>\n",
       "      <td>597.198547</td>\n",
       "      <td>Kirromycin_m796.409973_z1_rt618.98999</td>\n",
       "      <td>3.953642</td>\n",
       "      <td>1.193821e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12107588485736251567</th>\n",
       "      <td>1</td>\n",
       "      <td>379.755035</td>\n",
       "      <td>1423.827271</td>\n",
       "      <td>354.465881</td>\n",
       "      <td>406.498810</td>\n",
       "      <td>epemicinB_m1422.819946_z1_rt405.769989</td>\n",
       "      <td>4.597789</td>\n",
       "      <td>4.635282e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1595927192511345411</th>\n",
       "      <td>1</td>\n",
       "      <td>367.855164</td>\n",
       "      <td>1585.887329</td>\n",
       "      <td>343.602081</td>\n",
       "      <td>400.690826</td>\n",
       "      <td>epemicinA_m1584.880005_z1_rt386.850006</td>\n",
       "      <td>4.118316</td>\n",
       "      <td>5.144291e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      charge          RT           mz     RTstart       RTend  \\\n",
       "id                                                                              \n",
       "6540659878518515890        1   46.026489   402.207275   37.475689   54.632912   \n",
       "280489307960157370         1  103.041199   474.207275   91.933479  113.822960   \n",
       "9143164355464892809        1  265.116608   566.277283  243.804138  286.463776   \n",
       "10815806269762203077       1  373.222076   585.357300  353.652527  393.302399   \n",
       "14786621585738311964       1  286.942780   587.337280  275.987885  310.614807   \n",
       "10021272275670129286       1  298.578583   601.357300  259.177887  319.175507   \n",
       "9181399712805582150        2  379.947540   712.417297  333.265137  412.114746   \n",
       "11025134690498236929       2  367.807312   793.447266  326.792847  404.532959   \n",
       "12247998505257479685       1  563.042603   797.417297  537.345642  597.198547   \n",
       "12107588485736251567       1  379.755035  1423.827271  354.465881  406.498810   \n",
       "1595927192511345411        1  367.855164  1585.887329  343.602081  400.690826   \n",
       "\n",
       "                                                             PeptideRef  \\\n",
       "id                                                                        \n",
       "6540659878518515890               bioxalomycina2_m401.200012_z1_rt48.18   \n",
       "280489307960157370                   Unknown_m473.200012_z1_rt44.200001   \n",
       "9143164355464892809              Coelichelin_m565.27002_z1_rt268.049988   \n",
       "10815806269762203077   dehydroxynocardamine_m584.349976_z1_rt351.070007   \n",
       "14786621585738311964  desmethylenylnocardamine_m586.330017_z1_rt339....   \n",
       "10021272275670129286            desferrioxamineE_m600.349976_z1_rt329.5   \n",
       "9181399712805582150              epemicinB_m1422.819946_z2_rt405.769989   \n",
       "11025134690498236929             epemicinA_m1584.880005_z2_rt386.850006   \n",
       "12247998505257479685              Kirromycin_m796.409973_z1_rt618.98999   \n",
       "12107588485736251567             epemicinB_m1422.819946_z1_rt405.769989   \n",
       "1595927192511345411              epemicinA_m1584.880005_z1_rt386.850006   \n",
       "\n",
       "                        quality     intensity  \n",
       "id                                             \n",
       "6540659878518515890   -2.825665  2.926461e+04  \n",
       "280489307960157370    -0.763105  2.209964e+05  \n",
       "9143164355464892809   -7.000911  1.934900e+04  \n",
       "10815806269762203077  -7.667008  6.850766e+05  \n",
       "14786621585738311964  -7.992048  9.937172e+04  \n",
       "10021272275670129286 -10.806923  1.364715e+05  \n",
       "9181399712805582150    4.456927  9.244367e+09  \n",
       "11025134690498236929   4.391304  3.584970e+09  \n",
       "12247998505257479685   3.953642  1.193821e+06  \n",
       "12107588485736251567   4.597789  4.635282e+08  \n",
       "1595927192511345411    4.118316  5.144291e+07  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filename= './results/Epemicins/interim/FFMI_Epemicins.featureXML'\n",
    "fmap = FeatureMapDF()\n",
    "FeatureXMLFile().load(filename, fmap)\n",
    "DF= fmap.get_fmi_df()\n",
    "feature_csv= './results/Epemicins/FFMI_Epemicins.csv'\n",
    "DF.to_csv(feature_csv)\n",
    "display(DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "edde62aa2661007f0756e9790e7a328c288a583bf6ce768a355147dac67c8db8"
  },
  "kernelspec": {
   "display_name": "Python 3.9.4 64-bit (conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
