{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "15547c3c",
   "metadata": {},
   "source": [
    "## Convert files from Thermo\n",
    "\n",
    "ThermoRawFileParser https://github.com/compomics/ThermoRawFileParser\n",
    "\n",
    "Convert profile MS1 data from Thermo raw files to centroid (MS1 and MS2) mzml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f30ec3b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-11-17 17:01:55 INFO Started analyzing folder data/raw/\n",
      "2021-11-17 17:01:55 INFO The folder contains 4 RAW files\n",
      "2021-11-17 17:01:55 INFO Started parsing data/raw/20210927_POS_UMETAB222_55_ISP2_Plate-10_NBC_00734_rep1.raw\n",
      "2021-11-17 17:01:56 INFO Processing 4984 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-11-17 17:02:05 INFO Finished parsing data/raw/20210927_POS_UMETAB222_55_ISP2_Plate-10_NBC_00734_rep1.raw\n",
      "2021-11-17 17:02:05 INFO Started parsing data/raw/20210927_POS_UMETAB222_4_ISP2_Plate-10_NBC_00984_rep1.raw\n",
      "2021-11-17 17:02:05 INFO Processing 4620 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-11-17 17:02:13 INFO Finished parsing data/raw/20210927_POS_UMETAB222_4_ISP2_Plate-10_NBC_00984_rep1.raw\n",
      "2021-11-17 17:02:13 INFO Started parsing data/raw/20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_01014_rep1.raw\n",
      "2021-11-17 17:02:13 INFO Processing 5609 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-11-17 17:02:21 INFO Finished parsing data/raw/20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_01014_rep1.raw\n",
      "2021-11-17 17:02:21 INFO Started parsing data/raw/20210927_POS_UMETAB222_7_ISP2_Plate-10_NBC_01238_rep1.raw\n",
      "2021-11-17 17:02:22 INFO Processing 4972 MS scans\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "\n",
      "2021-11-17 17:02:30 INFO Finished parsing data/raw/20210927_POS_UMETAB222_7_ISP2_Plate-10_NBC_01238_rep1.raw\n"
     ]
    }
   ],
   "source": [
    "!mono resources/ThermoRawFileParser/ThermoRawFileParser.exe -d=\"data/raw/\" -o=\"results/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82179e30",
   "metadata": {},
   "source": [
    "##### Copy all the files to data/raw directory and you can use the following script to generate a samples.tsv file automatically:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cc98040f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>comment</th>\n",
       "      <th>MAPnumber</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20210827_UMETAB219_POS_DNPM_Plate-2_MDNAWGS14_...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20210827_UMETAB219_POS_ISP2_Plate-2_MDNAWGS14_...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20210827_UMETAB219_POS_ISP2_Plate-2_MDNAWGS14_...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20210827_UMETAB219_POS_ISP2_Plate-2_MDNAWGS14_...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20210927_POS_UMETAB222_4_ISP2_Plate-10_NBC_009...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20210927_POS_UMETAB222_55_ISP2_Plate-10_NBC_00...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>20210927_POS_UMETAB222_56_DNPM_Plate-10_NBC_00...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>20210927_POS_UMETAB222_57_FPY12_Plate-10_NBC_0...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>20210927_POS_UMETAB222_5_DNPM_Plate-10_NBC_009...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>20210927_POS_UMETAB222_6_FPY12_Plate-10_NBC_00...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_010...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21210924_UMETAB222_POS_4_ISP2_Plate-8_NBC_0052...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>21210924_UMETAB222_POS_5_DNPM_Plate-8_NBC_0052...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>21210924_UMETAB222_POS_6_FPY12_Plate-8_NBC_005...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          sample_name comment MAPnumber\n",
       "0   20210827_UMETAB219_POS_DNPM_Plate-2_MDNAWGS14_...                  \n",
       "1   20210827_UMETAB219_POS_ISP2_Plate-2_MDNAWGS14_...                  \n",
       "2   20210827_UMETAB219_POS_ISP2_Plate-2_MDNAWGS14_...                  \n",
       "3   20210827_UMETAB219_POS_ISP2_Plate-2_MDNAWGS14_...                  \n",
       "4   20210927_POS_UMETAB222_4_ISP2_Plate-10_NBC_009...                  \n",
       "5   20210927_POS_UMETAB222_55_ISP2_Plate-10_NBC_00...                  \n",
       "6   20210927_POS_UMETAB222_56_DNPM_Plate-10_NBC_00...                  \n",
       "7   20210927_POS_UMETAB222_57_FPY12_Plate-10_NBC_0...                  \n",
       "8   20210927_POS_UMETAB222_5_DNPM_Plate-10_NBC_009...                  \n",
       "9   20210927_POS_UMETAB222_6_FPY12_Plate-10_NBC_00...                  \n",
       "10  20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_010...                  \n",
       "11  21210924_UMETAB222_POS_4_ISP2_Plate-8_NBC_0052...                  \n",
       "12  21210924_UMETAB222_POS_5_DNPM_Plate-8_NBC_0052...                  \n",
       "13  21210924_UMETAB222_POS_6_FPY12_Plate-8_NBC_005...                  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!(cd data/raw && ls *.raw > filelist.txt)\n",
    "import pandas as pd\n",
    "header_list = [\"sample_name\"]\n",
    "df=pd.read_csv(\"data/raw/filelist.txt\", names=header_list, index_col= None)\n",
    "df[\"sample_name\"]=df[\"sample_name\"].replace(\".raw*\", value=\"\", regex=True)\n",
    "df[\"comment\"] = \" \"\n",
    "df[\"MAPnumber\"] = \" \"\n",
    "df.to_csv(\"config/samples.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7c0f75",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "969483bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyopenms import *\n",
    "from pandas import DataFrame\n",
    "import pandas as pd\n",
    "import pyteomics\n",
    "from pyteomics.openms import featurexml\n",
    "import numpy as np\n",
    "from pyteomics import mztab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da76ec1a",
   "metadata": {},
   "source": [
    "# Preprocessing step\n",
    "\n",
    "## The first preprocessing function consists of six steps:\n",
    "\n",
    "### 1) PrecursorCorrection (To the \"highest intensity MS1 peak\")\n",
    "\n",
    "This algorithm is used directly after the file introduction, in order to correct any wrong MS1 precursor annotation - this means it will correct the precursos m/z in a given mass range (e.g. 0.1 Da) and then use the highest intensity peak as the precursor. We assume that, in the given mass window, the precursor with the hightest intensity was actually fragmented (top-n method), which is a method used by default in the Thermo Orbitrap instrument (Center for Biosustainability).\n",
    "\n",
    "### OpenMS definition:\n",
    "\n",
    "\"Selection of the peak with the highest intensity as corrected precursor mass in a given mass range (e.g. precursor mass +/- 0.2 Da)\n",
    "\n",
    "For each MS2 spectrum the corresponding MS1 spectrum is determined by using the rt information of the precursor. In the MS1, the peak with the highest intensity in a given mass range to the uncorrected precursor m/z is selected and used as corrected precursor m/z.\"\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html//classOpenMS_1_1PrecursorCorrection.html#a8acf85ba8b9f249de0369bb083355982 \n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b538e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_filename= \"results/Epemicins/interim/Epemicins.mzML\" #introduce the mzML file that you want to preprocess here. If it is more than one, refer to the other jupyter notebook\n",
    "exp = MSExperiment()\n",
    "MzMLFile().load(example_filename, exp)\n",
    "exp.sortSpectra(True)\n",
    "    \n",
    "delta_mzs= []\n",
    "mzs = []\n",
    "rts= []\n",
    "PrecursorCorrection.correctToHighestIntensityMS1Peak(exp, 100.0, True, delta_mzs, mzs, rts)\n",
    "MzMLFile().store(\"results/Epemicins/interim/precursorcorrected_Epemicins.mzML\", exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b16c4b46",
   "metadata": {},
   "source": [
    "### 2) Mass trace detection\n",
    "\n",
    "A mass trace extraction method that gathers peaks similar in m/z and moving along retention time.\n",
    "\n",
    "Peaks of a MSExperiment are sorted by their intensity and stored in a list of potential chromatographic apex positions. Only peaks that are **above the noise threshold** (user-defined) are analyzed and only peaks that are n times above this minimal threshold are considered as apices. This saves computational resources and decreases the noise in the resulting output.\n",
    "\n",
    "Starting with these, mass traces are extended in retention time. During this extension phase, the centroid m/z is computed on-line as an intensity-weighted mean of peaks.\n",
    "\n",
    "The extension phase ends when either the frequency of gathered peaks drops below a threshold (min_sample_rate, see MassTraceDetection parameters) or when the number of missed scans exceeds a threshold (trace_termination_outliers, see MassTraceDetection parameters).\n",
    "\n",
    "Finally, only mass traces that pass a filter (a certain minimal and maximal length as well as having the minimal sample rate criterion fulfilled) get added to the result.\"\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1MassTraceDetection.html#abff6e392ce6da7af8f083397494a7971\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5f5c9697",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'mass trace detection':\n",
      "-- done [took 0.44 s (CPU), 0.43 s (Wall)] -- \n"
     ]
    }
   ],
   "source": [
    "mass_traces = []\n",
    "mtd = MassTraceDetection()\n",
    "mtd_par = mtd.getDefaults()\n",
    "mtd_par.setValue(\"mass_error_ppm\", 10.0) \n",
    "mtd_par.setValue(\"noise_threshold_int\", 1.0e04)\n",
    "mtd.setParameters(mtd_par)\n",
    "mtd.run(exp, mass_traces, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "561c3ab9",
   "metadata": {},
   "source": [
    "### 3) Elution peak detection\n",
    "\n",
    "\"Extracts chromatographic peaks from a mass trace.\n",
    "\n",
    "Mass traces may consist of several consecutively (partly overlapping) eluting peaks, e.g., stemming from (almost) isobaric compounds that are separated by retention time. Especially in metabolomics, isomeric compounds with exactly the same mass but different retentional behaviour may still be contained in the same mass trace.\n",
    "\n",
    "This method first applies smoothing on the mass trace's intensities, then detects local minima/maxima in order to separate the chromatographic peaks from each other. Detection of maxima is performed on the smoothed intensities and uses a fixed peak width (given as parameter chrom_fwhm) within which only a single maximum is expected. Currently smoothing is done using SavitzkyGolay smoothing with a second order polynomial and a frame length of the fixed peak width.\n",
    "\n",
    "Depending on the \"width_filtering\" parameters, mass traces are filtered by length in seconds (\"fixed\" filter) or by quantile.\n",
    "\n",
    "The output of the algorithm is a set of chromatographic peaks for each mass trace, i.e. a vector of split mass traces (see ElutionPeakDetection parameters).\n",
    "\n",
    "In general, a user would want to call the \"detectPeaks\" functions, potentially followed by the \"filterByPeakWidth\" function.\"\n",
    "\n",
    "This method in other words is \"deconvolution\".\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1ElutionPeakDetection.html#details\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "37d0c726",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'elution peak detection':\n",
      "-- done [took 2.18 s (CPU), 0.31 s (Wall)] -- \n"
     ]
    }
   ],
   "source": [
    "mass_traces_split = []\n",
    "mass_traces_final = []\n",
    "epd = ElutionPeakDetection()\n",
    "epd_par = epd.getDefaults()\n",
    "epd_par.setValue(\"width_filtering\", \"fixed\")\n",
    "epd.setParameters(epd_par)\n",
    "epd.detectPeaks(mass_traces, mass_traces_split)\n",
    "    \n",
    "if (epd.getParameters().getValue(\"width_filtering\") == \"auto\"):\n",
    "     epd.filterByPeakWidth(mass_traces_split, mass_traces_final)\n",
    "else:\n",
    "    mass_traces_final = mass_traces_split"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45e1dcab",
   "metadata": {},
   "source": [
    "### 4) Feature detection\n",
    "\n",
    "FeatureFinderMetabo assembles metabolite features from singleton mass traces.\n",
    "\n",
    "Mass traces alone would allow for further analysis such as metabolite ID or statistical evaluation. However, in general, monoisotopic mass traces are accompanied by satellite C13 peaks and thus may render the analysis more difficult. **FeatureFinderMetabo fulfills a further data reduction step by assembling compatible mass traces to metabolite features** (that is, all mass traces originating from one metabolite). To this end, multiple metabolite hypotheses are formulated and scored according to how well differences in RT (optional), m/z or intensity ratios match to those of theoretical isotope patterns.\n",
    "\n",
    "Detailed information is stored in the featureXML output: it contains meta-values for each feature about the mass trace differences (inspectable via TOPPView). \n",
    "\n",
    "By default, the linear model is used (better for small molecules - use the 13C for proteomics).\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_FeatureFinderMetabo.html\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c31f3921",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'assembling mass traces to features':\n",
      "-- done [took 4.40 s (CPU), 0.71 s (Wall)] -- \n"
     ]
    }
   ],
   "source": [
    "feature_map_FFM = FeatureMap()\n",
    "feat_chrom = []\n",
    "ffm = FeatureFindingMetabo()\n",
    "ffm_par = ffm.getDefaults() \n",
    "ffm_par.setValue(\"isotope_filtering_model\", \"none\")\n",
    "ffm_par.setValue(\"remove_single_traces\", \"false\")\n",
    "ffm_par.setValue(\"mz_scoring_by_elements\", \"false\")\n",
    "ffm_par.setValue(\"report_convex_hulls\", \"true\")\n",
    "ffm.setParameters(ffm_par)\n",
    "ffm.run(mass_traces_final, feature_map_FFM, feat_chrom)\n",
    "feature_map_FFM.setUniqueIds()\n",
    "fh = FeatureXMLFile()\n",
    "filename= \"results/Epemicins/interim/precursorcorrected_Epemicins.mzML\"\n",
    "feature_map_FFM.setPrimaryMSRunPath([filename.encode()])\n",
    "fh.store('./results/Epemicins/interim/FFM_Epemicins.featureXML', feature_map_FFM)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50d5ee28",
   "metadata": {},
   "source": [
    "### 5) Metabolite adduct decharger (MetaboliteFeatureDeconvolution)\n",
    "\n",
    "For each peak, this algorithm reconstructs neutral masses by enumerating all possible adducts with matching charge. Provide the probability of adduct formation according to your instrument using the parameter \"potential adducts\".\n",
    "\n",
    "With SIRIUS, you are only able to use singly charged adducts so charges higher than 1 are filtered out.\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/UTILS_MetaboliteAdductDecharger.html\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "59a21b2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MassExplainer table size: 12\n",
      "4611 spectra and 1 chromatograms stored.\n",
      "Generating Masses with threshold: -6.90776 ...\n",
      "done\n",
      "0 of 2852 valid net charge compomer results did not pass the feature charge constraints\n",
      "Inferring edges raised edge count from 2852 to 2852\n",
      "Found 2852 putative edges (of 131864) and avg hit-size of 1\n",
      "Using solver 'coinor' ...\n",
      "Optimal solution found!\n",
      "<Using solver 'coinor' ...> occurred 7 times\n",
      " Branch and cut took 0.919725 seconds,  with objective value: 6.88.\n",
      "<Optimal solution found!> occurred 7 times\n",
      "ILP score is: 6.88\n",
      "Agreeing charges: 541/3526\n",
      "ConsensusFeature::computeDechargeConsensus() WARNING: Feature's charge is 0! This will lead to M=0!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Compomer: Da -41.0003; q_net 0; logP -3.91202[[ (H2O1Na1) --> () ]]\n",
      " Compomer: Da -36.0449; q_net 0; logP -4.60517[[ (H2O1H4N1) --> () ]]\n",
      " Compomer: Da -21.9819; q_net 0; logP -2.12026[[ (Na1) --> (H1) ]]\n",
      " Compomer: Da -19.0184; q_net 0; logP -2.81341[[ (H1H2O1) --> () ]]\n",
      " Compomer: Da -17.0265; q_net 0; logP -2.81341[[ (H4N1) --> (H1) ]]\n",
      " Compomer: Da -4.9554; q_net 0; logP -3.91202[[ (Na1) --> (H4N1) ]]\n",
      " Compomer: Da 4.9554; q_net 0; logP -3.91202[[ (H4N1) --> (Na1) ]]\n",
      " Compomer: Da 17.0265; q_net 0; logP -2.81341[[ (H1) --> (H4N1) ]]\n",
      " Compomer: Da 19.0184; q_net 0; logP -2.81341[[ () --> (H1H2O1) ]]\n",
      " Compomer: Da 21.9819; q_net 0; logP -2.12026[[ (H1) --> (Na1) ]]\n",
      " Compomer: Da 36.0449; q_net 0; logP -4.60517[[ () --> (H2O1H4N1) ]]\n",
      " Compomer: Da 41.0003; q_net 0; logP -3.91202[[ () --> (H2O1Na1) ]]\n",
      " "
     ]
    }
   ],
   "source": [
    "mfd = MetaboliteFeatureDeconvolution()\n",
    "mdf_par = mfd.getDefaults()\n",
    "mdf_par.setValue(\"potential_adducts\",  [b\"H:+:0.6\",b\"Na:+:0.2\",b\"NH4:+:0.1\", b\"H2O:-:0.1\"])\n",
    "mdf_par.setValue(\"charge_min\", 1, \"Minimal possible charge\")\n",
    "mdf_par.setValue(\"charge_max\", 1, \"Maximal possible charge\")\n",
    "mdf_par.setValue(\"charge_span_max\", 1)\n",
    "mdf_par.setValue(\"max_neutrals\", 1)\n",
    "mfd.setParameters(mdf_par)\n",
    "    \n",
    "feature_map_DEC = FeatureMap()\n",
    "cons_map0 = ConsensusMap()\n",
    "cons_map1 = ConsensusMap()\n",
    "mfd.compute(feature_map_FFM, feature_map_DEC, cons_map0, cons_map1)\n",
    "fxml = FeatureXMLFile()\n",
    "fh.store('./results/Epemicins/interim/MFD_Epemicins.featureXML', feature_map_DEC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99c2993c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from functools import reduce\n",
    "from pathlib import Path\n",
    "from time import perf_counter\n",
    "import sys\n",
    "\n",
    "from IPython.core.display import display\n",
    "from pandas import CategoricalDtype\n",
    "import numpy as np\n",
    "from pyopenms import *\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "class FeatureMapDF(FeatureMap):      \n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "\n",
    "    def get_df(self):\n",
    "        def gen(fmap: FeatureMap, fun):\n",
    "            for f in fmap:\n",
    "                yield from fun(f)\n",
    "\n",
    "        def extractMetaData(f: Feature):\n",
    "            metavals = []\n",
    "            types = []\n",
    "            for k in metavals:\n",
    "                if k == b\"num_of_masstraces\":\n",
    "                    types.append('f')\n",
    "                elif k== b\"FWHM\":\n",
    "                    types.append('f')\n",
    "                elif k== b\"max_height\":\n",
    "                    types.append('i4')\n",
    "                elif k== b\"masstrace_intensity\":\n",
    "                    types.append(\"f\")\n",
    "            # subfeatures = f.getFeatureList()  # type: list[FeatureHandle]\n",
    "            pep = f.getPeptideIdentifications()  # type: list[PeptideIdentification]\n",
    "            bb = f.getConvexHull().getBoundingBox2D()\n",
    "            for u in metavals:\n",
    "                f.getMetaValue(u)\n",
    "            if len(pep) != 0:\n",
    "                hits = pep[0].getHits()\n",
    "                if len(hits) != 0:\n",
    "                    besthit = hits[0]  # type: PeptideHit\n",
    "                    yield f.getUniqueId(), besthit.getSequence().toString(), f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"num_of_masstraces\"), f.getOverallQuality(), f.getIntensity()\n",
    "                else:\n",
    "                    yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"num_of_masstraces\"), f.getOverallQuality(), f.getIntensity()\n",
    "            else:\n",
    "                yield f.getUniqueId(), None, f.getCharge(), f.getRT(), f.getMZ(), bb[0][0], bb[1][0], f.getMetaValue(\"num_of_masstraces\"), f.getOverallQuality(), f.getIntensity()\n",
    "\n",
    "        cnt = self.size()\n",
    "\n",
    "        mddtypes = [('id', np.dtype('uint64')), ('sequence', 'U200'), ('charge', 'i4'), ('RT', 'f'), ('mz', 'f'),\n",
    "                    ('RTstart', 'f'), ('RTend', 'f'), (\"num_of_masstraces\", 'f'),\n",
    "                    ('quality', 'f'), ('intensity', 'f')]\n",
    "        mdarr = np.fromiter(iter=gen(self, extractMetaData), dtype=mddtypes, count=cnt)\n",
    "        df= pd.DataFrame(mdarr).set_index('id').sort_values(\"mz\").drop(columns= \"sequence\")\n",
    "        df= df[df[\"num_of_masstraces\"]>=2]\n",
    "        return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50aeae13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>charge</th>\n",
       "      <th>RT</th>\n",
       "      <th>mz</th>\n",
       "      <th>RTstart</th>\n",
       "      <th>RTend</th>\n",
       "      <th>num_of_masstraces</th>\n",
       "      <th>quality</th>\n",
       "      <th>intensity</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15695211054020622558</th>\n",
       "      <td>1</td>\n",
       "      <td>329.607452</td>\n",
       "      <td>208.133240</td>\n",
       "      <td>326.792847</td>\n",
       "      <td>335.102264</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>3.212667e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11553075680875761771</th>\n",
       "      <td>1</td>\n",
       "      <td>310.614807</td>\n",
       "      <td>208.133255</td>\n",
       "      <td>303.933105</td>\n",
       "      <td>315.405701</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>1.985221e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16931706070924112617</th>\n",
       "      <td>1</td>\n",
       "      <td>379.585754</td>\n",
       "      <td>260.185669</td>\n",
       "      <td>362.698792</td>\n",
       "      <td>455.897400</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>4.589339e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464053971863077264</th>\n",
       "      <td>1</td>\n",
       "      <td>228.511551</td>\n",
       "      <td>261.123413</td>\n",
       "      <td>222.785919</td>\n",
       "      <td>235.216217</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>4.301382e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14207239120303140199</th>\n",
       "      <td>1</td>\n",
       "      <td>316.367584</td>\n",
       "      <td>271.060028</td>\n",
       "      <td>302.927643</td>\n",
       "      <td>321.092896</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>1.677704e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13812678751659148901</th>\n",
       "      <td>1</td>\n",
       "      <td>350.433289</td>\n",
       "      <td>1603.895020</td>\n",
       "      <td>348.765778</td>\n",
       "      <td>355.269623</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.001192</td>\n",
       "      <td>1.190649e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4289608483695663489</th>\n",
       "      <td>1</td>\n",
       "      <td>342.791321</td>\n",
       "      <td>1604.396851</td>\n",
       "      <td>337.522614</td>\n",
       "      <td>352.043243</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>4.679359e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7662137660621363607</th>\n",
       "      <td>2</td>\n",
       "      <td>47.941818</td>\n",
       "      <td>1639.049805</td>\n",
       "      <td>43.264481</td>\n",
       "      <td>61.428474</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>2.294459e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7650966078180970882</th>\n",
       "      <td>1</td>\n",
       "      <td>173.429962</td>\n",
       "      <td>1646.474365</td>\n",
       "      <td>169.596756</td>\n",
       "      <td>181.128769</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>5.192035e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12543873065339080715</th>\n",
       "      <td>2</td>\n",
       "      <td>105.360298</td>\n",
       "      <td>1775.552734</td>\n",
       "      <td>101.564934</td>\n",
       "      <td>115.742714</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>1.904661e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2428 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      charge          RT           mz     RTstart       RTend  \\\n",
       "id                                                                              \n",
       "15695211054020622558       1  329.607452   208.133240  326.792847  335.102264   \n",
       "11553075680875761771       1  310.614807   208.133255  303.933105  315.405701   \n",
       "16931706070924112617       1  379.585754   260.185669  362.698792  455.897400   \n",
       "2464053971863077264        1  228.511551   261.123413  222.785919  235.216217   \n",
       "14207239120303140199       1  316.367584   271.060028  302.927643  321.092896   \n",
       "...                      ...         ...          ...         ...         ...   \n",
       "13812678751659148901       1  350.433289  1603.895020  348.765778  355.269623   \n",
       "4289608483695663489        1  342.791321  1604.396851  337.522614  352.043243   \n",
       "7662137660621363607        2   47.941818  1639.049805   43.264481   61.428474   \n",
       "7650966078180970882        1  173.429962  1646.474365  169.596756  181.128769   \n",
       "12543873065339080715       2  105.360298  1775.552734  101.564934  115.742714   \n",
       "\n",
       "                      num_of_masstraces   quality     intensity  \n",
       "id                                                               \n",
       "15695211054020622558                2.0  0.000180  3.212667e+07  \n",
       "11553075680875761771                2.0  0.000011  1.985221e+06  \n",
       "16931706070924112617                2.0  0.000258  4.589339e+07  \n",
       "2464053971863077264                 2.0  0.000025  4.301382e+06  \n",
       "14207239120303140199                2.0  0.000009  1.677704e+06  \n",
       "...                                 ...       ...           ...  \n",
       "13812678751659148901                3.0  0.001192  1.190649e+08  \n",
       "4289608483695663489                 2.0  0.000371  4.679359e+07  \n",
       "7662137660621363607                 2.0  0.000003  2.294459e+05  \n",
       "7650966078180970882                 3.0  0.000006  5.192035e+05  \n",
       "12543873065339080715                6.0  0.000049  1.904661e+06  \n",
       "\n",
       "[2428 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filename= './results/Epemicins/interim/MFD_Epemicins.featureXML'\n",
    "fmap = FeatureMapDF()\n",
    "FeatureXMLFile().load(filename, fmap)\n",
    "DF= fmap.get_df()\n",
    "feature_csv= './results/Epemicins/features_Epemicins.csv'\n",
    "DF.to_csv(feature_csv)\n",
    "display(DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14b8a4ae",
   "metadata": {},
   "source": [
    "### 6) PrecursorCorrection (To the \"nearest feature”)\n",
    "\n",
    "This algorithm is used after feature detection, adduct grouping and even identification via accurate mass search. It basically allows the precursor correction on MS2 level. \n",
    "\n",
    "Which means that if there are MS2 spectra in my feature space which have been measured in isotope traces, it “corrects” the MS2 spectrum annotation to the monoisotopic trace. That is why you have a high mass deviation 100 pm, but 0.0 rt tolerance. So it basically corrects the MS2 to the feature centroid that can be found/mapped by SIRIUS::preprocessing later on.\n",
    "\n",
    "Wrong assignment of the mono-isotopic mass for precursors are assumed:\n",
    "\n",
    "1. if precursor_mz matches the mz of a non-monoisotopic feature mass trace and \n",
    "2. in the case that believe_charge is true: if feature_charge matches the precursor_charge In the case of wrong mono-isotopic assignment several options for correction are available: keep_original will create a copy of the precursor and tandem spectrum for the new mono-isotopic mass trace and retain the original one. all_matching_features does this not for only the closest feature but all features in a question.\n",
    "\n",
    "#### Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1PrecursorCorrection.html\n",
    "\n",
    "###### Citation: Röst, H.L., Sachsenberg, T., Aiche, S., Bielow, C., Weisser, H., Aicheler, F., Andreotti, S., Ehrlich, H.-C., Gutenbrunner, P., Kenar, E., Liang, X., Nahnsen, S., Nilse, L., Pfeuffer, J., Rosenberger, G., Rurik, M., Schmitt, U., Veit, J., Walzer, M., Wojnar, D., Wolski, W.E.,Schilling, O., Choudhary, J.S., Malmström, L., Aebersold, R., Reinert, K., Kohlbacher, O. OpenMS: A flexible open-source software platform for mass spectrometry data analysis. Nature Methods, vol. 13, 2016. doi:10.1038/nmeth.3959"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e793c80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HighResPrecursorMassCorrector warning: at least one feature has no convex hull - omitting feature for matching\n"
     ]
    }
   ],
   "source": [
    "PrecursorCorrection.correctToNearestFeature(feature_map_DEC, exp, 0.0, 100.0, True, False, False, False, 3, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26bd9800",
   "metadata": {},
   "source": [
    "### SIRIUS Adapter\n",
    "The SIRIUS function is optional and includes the SIRIUS Adapter Algorithm from the Boecher lab. \n",
    "\n",
    "\n",
    "The algorithm generates formula prediction from scores calculated from 1) MS2 fragmentation scores (ppm error + intensity) and 2) MS1 isotopic pattern scores.\n",
    "\n",
    "\n",
    "It can only compute feautures that are singly charged. There is also a timeout for compounds (compound timeout so that it doesn't compute for longer than 100 seconds per feature, which normally happens with larger molecules).\n",
    "\n",
    "\n",
    "-sirius:compound_timeout <number>                    \n",
    "Maximal computation time in seconds for a single compound. 0 for an infinite amount of time. (default: '100' min: '0')\n",
    "\n",
    "    \n",
    "This algorith can help in data dereplication and analysis for direct library search. \n",
    "\n",
    "#### Documentation: https://boecker-lab.github.io/docs.sirius.github.io/\n",
    "\n",
    "###### Citation: Kai Dührkop, Markus Fleischauer, Marcus Ludwig, Alexander A. Aksenov, Alexey V. Melnik, Marvin Meusel, Pieter C. Dorrestein, Juho Rousu, and Sebastian Böcker, SIRIUS 4: Turning tandem mass spectra into metabolite structure information. Nature Methods 16, 299–302, 2019."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e2bd60",
   "metadata": {},
   "source": [
    "We need to preprocess (FeatureFinderMetabo) again without convex hulls this time, otherwise the results from SIRIUS are incorrect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cc2c4e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_map_FFM = FeatureMap()\n",
    "feat_chrom = []\n",
    "ffm = FeatureFindingMetabo()\n",
    "ffm_par = ffm.getDefaults() \n",
    "ffm_par.setValue(\"isotope_filtering_model\", \"none\")\n",
    "ffm_par.setValue(\"remove_single_traces\", \"false\")\n",
    "ffm_par.setValue(\"mz_scoring_by_elements\", \"false\")\n",
    "ffm.setParameters(ffm_par)\n",
    "ffm.run(mass_traces_final, feature_map_FFM, feat_chrom)\n",
    "feature_map_FFM.setUniqueIds()\n",
    "fh = FeatureXMLFile()\n",
    "filename= \"results/Epemicins/interim/precursorcorrected_Epemicins.mzML\"\n",
    "feature_map_FFM.setPrimaryMSRunPath([filename.encode()])\n",
    "fh.store('./results/Epemicins/interim/FFM_noconvexhulls_Epemicins.featureXML', feature_map_FFM)\n",
    "mfd = MetaboliteFeatureDeconvolution()\n",
    "mdf_par = mfd.getDefaults()\n",
    "mdf_par.setValue(\"potential_adducts\",  [b\"H:+:0.6\",b\"Na:+:0.2\",b\"NH4:+:0.1\", b\"H2O:-:0.1\"])\n",
    "mdf_par.setValue(\"charge_min\", 1, \"Minimal possible charge\")\n",
    "mdf_par.setValue(\"charge_max\", 1, \"Maximal possible charge\")\n",
    "mdf_par.setValue(\"charge_span_max\", 1)\n",
    "mdf_par.setValue(\"max_neutrals\", 1)\n",
    "mfd.setParameters(mdf_par)\n",
    "feature_map_DEC = FeatureMap()\n",
    "cons_map0 = ConsensusMap()\n",
    "cons_map1 = ConsensusMap()\n",
    "mfd.compute(feature_map_FFM, feature_map_DEC, cons_map0, cons_map1)\n",
    "fxml = FeatureXMLFile()\n",
    "fh.store('./results/Epemicins/interim/MFD_noconvexhulls_Epemicins.featureXML', feature_map_DEC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38c19764",
   "metadata": {},
   "outputs": [],
   "source": [
    "sirius_algo = SiriusAdapterAlgorithm()\n",
    "sirius_algo_par = sirius_algo.getDefaults()\n",
    "sirius_algo_par.setValue(\"preprocessing:filter_by_num_masstraces\", 2) #Number of mass traces each feature has to have to be included\n",
    "sirius_algo_par.setValue(\"preprocessing:feature_only\", \"true\") #Uses the feature information from in_featureinfo to reduce the search space to MS2\n",
    "sirius_algo_par.setValue(\"sirius:profile\", \"orbitrap\")\n",
    "sirius_algo_par.setValue(\"sirius:db\", \"none\")\n",
    "sirius_algo_par.setValue(\"sirius:ions_considered\", \"[M+H]+, [M-H2O+H]+, [M+Na]+, [M+NH4]+\")\n",
    "sirius_algo_par.setValue(\"sirius:elements_enforced\", \"CHN[15]OS[4]\") \n",
    "sirius_algo_par.setValue(\"project:processors\", 2)\n",
    "sirius_algo_par.setValue(\"fingerid:db\", \"BIO\")\n",
    "sirius_algo.setParameters(sirius_algo_par)\n",
    "    \n",
    "featureinfo = \"./results/Epemicins/interim/MFD_noconvexhulls_Epemicins.featureXML\"\n",
    "fm_info = FeatureMapping_FeatureMappingInfo()\n",
    "feature_mapping = FeatureMapping_FeatureToMs2Indices() \n",
    "sirius_algo.preprocessingSirius(featureinfo,\n",
    "                                exp,\n",
    "                                fm_info,\n",
    "                                feature_mapping)\n",
    "sirius_algo.logFeatureSpectraNumber(featureinfo, \n",
    "                                    feature_mapping,\n",
    "                                    exp)\n",
    "msfile = SiriusMSFile()\n",
    "debug_level = 3\n",
    "sirius_tmp = SiriusTemporaryFileSystemObjects(debug_level)\n",
    "siriusstring= String(sirius_tmp.getTmpMsFile())\n",
    "feature_only = sirius_algo.isFeatureOnly()\n",
    "isotope_pattern_iterations = sirius_algo.getIsotopePatternIterations()\n",
    "no_mt_info = sirius_algo.isNoMasstraceInfoIsotopePattern()\n",
    "compound_info = []\n",
    "msfile.store(exp,\n",
    "             String(sirius_tmp.getTmpMsFile()),\n",
    "             feature_mapping, \n",
    "             feature_only,\n",
    "             isotope_pattern_iterations, \n",
    "             no_mt_info, \n",
    "             compound_info)\n",
    "out_csifingerid = \"./results/Epemicins/interim/structures_Epemicins.mzTab\" \n",
    "executable= \"/path/to/sirius_executable\"\n",
    "subdirs = sirius_algo.callSiriusQProcess(String(sirius_tmp.getTmpMsFile()),\n",
    "                                         String(sirius_tmp.getTmpOutDir()),\n",
    "                                         String(executable),\n",
    "                                         String(out_csifingerid),\n",
    "                                         False)\n",
    "candidates = sirius_algo.getNumberOfSiriusCandidates()\n",
    "sirius_result = MzTab()\n",
    "siriusfile = MzTabFile()\n",
    "SiriusMzTabWriter.read(subdirs,\n",
    "                       filename,\n",
    "                       candidates,\n",
    "                       sirius_result)\n",
    "siriusfile.store(\"./results/Epemicins/interim/formulas_Epemicins.mzTab\", sirius_result) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5ffce560",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyteomics\n",
    "from pyteomics import mztab\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3d00a50f-7f8b-4e37-b06b-e25af34de8f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chemical_formula</th>\n",
       "      <th>exp_mass_to_charge</th>\n",
       "      <th>retention_time</th>\n",
       "      <th>SiriusScore</th>\n",
       "      <th>TreeScore</th>\n",
       "      <th>IsotopeScore</th>\n",
       "      <th>opt_global_adduct</th>\n",
       "      <th>opt_gobal_precursorFormula</th>\n",
       "      <th>opt_global_rank</th>\n",
       "      <th>opt_global_explainedPeaks</th>\n",
       "      <th>opt_global_explainedIntensity</th>\n",
       "      <th>opt_global_median_mass_error_fragment_peaks_ppm</th>\n",
       "      <th>opt_global_median_absolute_mass_error_fragment_peaks_ppm</th>\n",
       "      <th>opt_global_mass_error_precursor_ppm</th>\n",
       "      <th>opt_global_compoundId</th>\n",
       "      <th>opt_global_compoundScanNumber</th>\n",
       "      <th>opt_global_featureId</th>\n",
       "      <th>opt_global_native_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>C5H11NO4</td>\n",
       "      <td>150.077119</td>\n",
       "      <td>61.200087</td>\n",
       "      <td>30.714977</td>\n",
       "      <td>29.001456</td>\n",
       "      <td>1.713521</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>C5H11NO4</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>0.476666</td>\n",
       "      <td>6.283649</td>\n",
       "      <td>8.130967</td>\n",
       "      <td>6.896623</td>\n",
       "      <td>164</td>\n",
       "      <td>165</td>\n",
       "      <td>id_8492465156883061367</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=165|c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>C6H11N2O</td>\n",
       "      <td>150.077119</td>\n",
       "      <td>61.200087</td>\n",
       "      <td>66.193542</td>\n",
       "      <td>63.986303</td>\n",
       "      <td>2.207239</td>\n",
       "      <td>[M + Na]+</td>\n",
       "      <td>C6H11N2O</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.699695</td>\n",
       "      <td>9.998139</td>\n",
       "      <td>9.998139</td>\n",
       "      <td>5.068310</td>\n",
       "      <td>164</td>\n",
       "      <td>165</td>\n",
       "      <td>id_8492465156883061367</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=165|c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>C8H9N2O</td>\n",
       "      <td>150.077119</td>\n",
       "      <td>61.200087</td>\n",
       "      <td>60.457524</td>\n",
       "      <td>58.762027</td>\n",
       "      <td>1.695497</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>C8H9N2O</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>0.690505</td>\n",
       "      <td>-10.961369</td>\n",
       "      <td>10.961369</td>\n",
       "      <td>-10.961369</td>\n",
       "      <td>164</td>\n",
       "      <td>165</td>\n",
       "      <td>id_8492465156883061367</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=165|c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>C6H7N5</td>\n",
       "      <td>150.077119</td>\n",
       "      <td>61.200087</td>\n",
       "      <td>57.863930</td>\n",
       "      <td>55.676457</td>\n",
       "      <td>2.187473</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>C6H7N5</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0.690505</td>\n",
       "      <td>-1.036984</td>\n",
       "      <td>1.327234</td>\n",
       "      <td>-2.014742</td>\n",
       "      <td>164</td>\n",
       "      <td>165</td>\n",
       "      <td>id_8492465156883061367</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=165|c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>C6H4N4</td>\n",
       "      <td>150.077119</td>\n",
       "      <td>61.200087</td>\n",
       "      <td>46.961272</td>\n",
       "      <td>44.773799</td>\n",
       "      <td>2.187473</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>C6H8N4</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>0.665468</td>\n",
       "      <td>-0.896564</td>\n",
       "      <td>1.207026</td>\n",
       "      <td>-2.014742</td>\n",
       "      <td>164</td>\n",
       "      <td>165</td>\n",
       "      <td>id_8492465156883061367</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=165|c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1615</th>\n",
       "      <td>C45H65N15O15S</td>\n",
       "      <td>1110.441831</td>\n",
       "      <td>371.585916</td>\n",
       "      <td>52.606908</td>\n",
       "      <td>46.113926</td>\n",
       "      <td>6.492982</td>\n",
       "      <td>[M + Na]+</td>\n",
       "      <td>C45H65N15O15S</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.734734</td>\n",
       "      <td>-2.120817</td>\n",
       "      <td>2.349533</td>\n",
       "      <td>1.875550</td>\n",
       "      <td>2962</td>\n",
       "      <td>2963</td>\n",
       "      <td>id_6451041861266995604</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1616</th>\n",
       "      <td>C49H69N9O17S</td>\n",
       "      <td>1110.441831</td>\n",
       "      <td>371.585916</td>\n",
       "      <td>52.002975</td>\n",
       "      <td>46.115658</td>\n",
       "      <td>5.887317</td>\n",
       "      <td>[M + Na]+</td>\n",
       "      <td>C49H69N9O17S</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>0.734734</td>\n",
       "      <td>-2.120817</td>\n",
       "      <td>2.349533</td>\n",
       "      <td>-0.542737</td>\n",
       "      <td>2962</td>\n",
       "      <td>2963</td>\n",
       "      <td>id_6451041861266995604</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1617</th>\n",
       "      <td>C46H65N13O18</td>\n",
       "      <td>1110.441831</td>\n",
       "      <td>371.585916</td>\n",
       "      <td>49.912802</td>\n",
       "      <td>46.484328</td>\n",
       "      <td>3.428474</td>\n",
       "      <td>[M + Na]+</td>\n",
       "      <td>C46H65N13O18</td>\n",
       "      <td>3</td>\n",
       "      <td>30</td>\n",
       "      <td>0.743992</td>\n",
       "      <td>-2.558511</td>\n",
       "      <td>2.711593</td>\n",
       "      <td>-4.000363</td>\n",
       "      <td>2962</td>\n",
       "      <td>2963</td>\n",
       "      <td>id_6451041861266995604</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1619</th>\n",
       "      <td>C51H69N5O21</td>\n",
       "      <td>1110.441831</td>\n",
       "      <td>371.585916</td>\n",
       "      <td>49.447782</td>\n",
       "      <td>44.861487</td>\n",
       "      <td>4.586295</td>\n",
       "      <td>[M + Na]+</td>\n",
       "      <td>C51H69N5O21</td>\n",
       "      <td>5</td>\n",
       "      <td>29</td>\n",
       "      <td>0.742894</td>\n",
       "      <td>-1.862543</td>\n",
       "      <td>2.403428</td>\n",
       "      <td>3.697485</td>\n",
       "      <td>2962</td>\n",
       "      <td>2963</td>\n",
       "      <td>id_6451041861266995604</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1618</th>\n",
       "      <td>C47H65N11O19</td>\n",
       "      <td>1110.441831</td>\n",
       "      <td>371.585916</td>\n",
       "      <td>49.661430</td>\n",
       "      <td>45.969773</td>\n",
       "      <td>3.691657</td>\n",
       "      <td>[M + Na]+</td>\n",
       "      <td>C47H65N11O19</td>\n",
       "      <td>4</td>\n",
       "      <td>29</td>\n",
       "      <td>0.723546</td>\n",
       "      <td>-1.067860</td>\n",
       "      <td>2.822854</td>\n",
       "      <td>6.115773</td>\n",
       "      <td>2962</td>\n",
       "      <td>2963</td>\n",
       "      <td>id_6451041861266995604</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1769 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chemical_formula  exp_mass_to_charge  retention_time  SiriusScore  \\\n",
       "58           C5H11NO4          150.077119       61.200087    30.714977   \n",
       "54           C6H11N2O          150.077119       61.200087    66.193542   \n",
       "55            C8H9N2O          150.077119       61.200087    60.457524   \n",
       "56             C6H7N5          150.077119       61.200087    57.863930   \n",
       "57             C6H4N4          150.077119       61.200087    46.961272   \n",
       "...               ...                 ...             ...          ...   \n",
       "1615    C45H65N15O15S         1110.441831      371.585916    52.606908   \n",
       "1616     C49H69N9O17S         1110.441831      371.585916    52.002975   \n",
       "1617     C46H65N13O18         1110.441831      371.585916    49.912802   \n",
       "1619      C51H69N5O21         1110.441831      371.585916    49.447782   \n",
       "1618     C47H65N11O19         1110.441831      371.585916    49.661430   \n",
       "\n",
       "      TreeScore  IsotopeScore opt_global_adduct opt_gobal_precursorFormula  \\\n",
       "58    29.001456      1.713521          [M + H]+                   C5H11NO4   \n",
       "54    63.986303      2.207239         [M + Na]+                   C6H11N2O   \n",
       "55    58.762027      1.695497          [M + H]+                    C8H9N2O   \n",
       "56    55.676457      2.187473          [M + H]+                     C6H7N5   \n",
       "57    44.773799      2.187473     [M + H4 + N]+                     C6H8N4   \n",
       "...         ...           ...               ...                        ...   \n",
       "1615  46.113926      6.492982         [M + Na]+              C45H65N15O15S   \n",
       "1616  46.115658      5.887317         [M + Na]+               C49H69N9O17S   \n",
       "1617  46.484328      3.428474         [M + Na]+               C46H65N13O18   \n",
       "1619  44.861487      4.586295         [M + Na]+                C51H69N5O21   \n",
       "1618  45.969773      3.691657         [M + Na]+               C47H65N11O19   \n",
       "\n",
       "      opt_global_rank  opt_global_explainedPeaks  \\\n",
       "58                  5                         10   \n",
       "54                  1                         16   \n",
       "55                  2                         15   \n",
       "56                  3                         15   \n",
       "57                  4                         11   \n",
       "...               ...                        ...   \n",
       "1615                1                         30   \n",
       "1616                2                         30   \n",
       "1617                3                         30   \n",
       "1619                5                         29   \n",
       "1618                4                         29   \n",
       "\n",
       "      opt_global_explainedIntensity  \\\n",
       "58                         0.476666   \n",
       "54                         0.699695   \n",
       "55                         0.690505   \n",
       "56                         0.690505   \n",
       "57                         0.665468   \n",
       "...                             ...   \n",
       "1615                       0.734734   \n",
       "1616                       0.734734   \n",
       "1617                       0.743992   \n",
       "1619                       0.742894   \n",
       "1618                       0.723546   \n",
       "\n",
       "      opt_global_median_mass_error_fragment_peaks_ppm  \\\n",
       "58                                           6.283649   \n",
       "54                                           9.998139   \n",
       "55                                         -10.961369   \n",
       "56                                          -1.036984   \n",
       "57                                          -0.896564   \n",
       "...                                               ...   \n",
       "1615                                        -2.120817   \n",
       "1616                                        -2.120817   \n",
       "1617                                        -2.558511   \n",
       "1619                                        -1.862543   \n",
       "1618                                        -1.067860   \n",
       "\n",
       "      opt_global_median_absolute_mass_error_fragment_peaks_ppm  \\\n",
       "58                                             8.130967          \n",
       "54                                             9.998139          \n",
       "55                                            10.961369          \n",
       "56                                             1.327234          \n",
       "57                                             1.207026          \n",
       "...                                                 ...          \n",
       "1615                                           2.349533          \n",
       "1616                                           2.349533          \n",
       "1617                                           2.711593          \n",
       "1619                                           2.403428          \n",
       "1618                                           2.822854          \n",
       "\n",
       "      opt_global_mass_error_precursor_ppm  opt_global_compoundId  \\\n",
       "58                               6.896623                    164   \n",
       "54                               5.068310                    164   \n",
       "55                             -10.961369                    164   \n",
       "56                              -2.014742                    164   \n",
       "57                              -2.014742                    164   \n",
       "...                                   ...                    ...   \n",
       "1615                             1.875550                   2962   \n",
       "1616                            -0.542737                   2962   \n",
       "1617                            -4.000363                   2962   \n",
       "1619                             3.697485                   2962   \n",
       "1618                             6.115773                   2962   \n",
       "\n",
       "      opt_global_compoundScanNumber    opt_global_featureId  \\\n",
       "58                              165  id_8492465156883061367   \n",
       "54                              165  id_8492465156883061367   \n",
       "55                              165  id_8492465156883061367   \n",
       "56                              165  id_8492465156883061367   \n",
       "57                              165  id_8492465156883061367   \n",
       "...                             ...                     ...   \n",
       "1615                           2963  id_6451041861266995604   \n",
       "1616                           2963  id_6451041861266995604   \n",
       "1617                           2963  id_6451041861266995604   \n",
       "1619                           2963  id_6451041861266995604   \n",
       "1618                           2963  id_6451041861266995604   \n",
       "\n",
       "                                   opt_global_native_id  \n",
       "58    controllerType=0 controllerNumber=1 scan=165|c...  \n",
       "54    controllerType=0 controllerNumber=1 scan=165|c...  \n",
       "55    controllerType=0 controllerNumber=1 scan=165|c...  \n",
       "56    controllerType=0 controllerNumber=1 scan=165|c...  \n",
       "57    controllerType=0 controllerNumber=1 scan=165|c...  \n",
       "...                                                 ...  \n",
       "1615      controllerType=0 controllerNumber=1 scan=2963  \n",
       "1616      controllerType=0 controllerNumber=1 scan=2963  \n",
       "1617      controllerType=0 controllerNumber=1 scan=2963  \n",
       "1619      controllerType=0 controllerNumber=1 scan=2963  \n",
       "1618      controllerType=0 controllerNumber=1 scan=2963  \n",
       "\n",
       "[1769 rows x 18 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "siriusfile= \"./results/20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_01014_rep1/interim/sirius/formulas_20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_01014_rep1.mzTab\"\n",
    "sirius=  pyteomics.mztab.MzTab(siriusfile, encoding='UTF8', table_format='df')\n",
    "sirius.metadata\n",
    "df= sirius.small_molecule_table\n",
    "SIRIUS_DF= df.drop(columns= [\"identifier\", \"smiles\", \"inchi_key\", \"description\", \"calc_mass_to_charge\", \"charge\", \"taxid\", \"species\",\"database\", \"database_version\", \"spectra_ref\", \"search_engine\", \"modifications\"])\n",
    "SIRIUS_DF=SIRIUS_DF[SIRIUS_DF[\"opt_global_explainedIntensity\"] >= 0.4] #opt_global_explainedIntensity should be higher than 0.8 or 0.9 even for reliable results. We filter out only the ones below 0.4 since there are cases of correct formula assignment in the range above that number\n",
    "SIRIUS_DF= SIRIUS_DF.sort_values(by= \"exp_mass_to_charge\")\n",
    "SIRIUS_DF= SIRIUS_DF.rename(columns= {\"best_search_engine_score[1]\":\t\"SiriusScore\"}) #This score is the sum of the fragmentation pattern scoring (\"TreeScore\") and the isotope pattern scoring (\"IsotopeScore\")\n",
    "SIRIUS_DF= SIRIUS_DF.rename(columns= {\"best_search_engine_score[2]\":\t\"TreeScore\"})\n",
    "SIRIUS_DF= SIRIUS_DF.rename(columns= {\"best_search_engine_score[3]\":\t\"IsotopeScore\"}) #The closer to 10, the higher the quality of the isotope pattern\n",
    "SIRIUS_DF=SIRIUS_DF[SIRIUS_DF[\"IsotopeScore\"] > 0.0] #Isotope scores that are equal to zero are very low quality and filtered out\n",
    "SIRIUS_DF.to_csv(\"./results/20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_01014_rep1/interim/sirius/formulas_20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_01014_rep1.csv\")\n",
    "SIRIUS_DF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d4c0d4e-ffc1-4865-a9e5-129f3d3c0592",
   "metadata": {},
   "source": [
    "### CSIFinger:ID\n",
    "\n",
    "The CSI_fingerID function is another algorithm from the Boecher lab, just like SIRIUS adapter and is using the formula predictions from SIRIUS, to search in structural libraries and predict the structure of each formula\n",
    "\n",
    "#### Documentation: https://boecker-lab.github.io/docs.sirius.github.io/\n",
    "\n",
    "##### Citation: Kai Dührkop, Huibin Shen, Marvin Meusel, Juho Rousu, and Sebastian Böcker, Searching molecular structure databases with tandem mass spectra using CSI:FingerID, PNAS October 13, 2015 112 (41) 12580-12585, https://doi.org/10.1073/pnas.1509788112"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "261217e8-39a3-4ae4-9e22-18b110645b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "csi_file.store(\"./results/Epemicins/interim/structures_Epemicins.mzTab\", csi_result)\n",
    "csi_file= \"./results/Epemicins/interim/structures_Epemicins.mzTab\"\n",
    "CSI=  pyteomics.mztab.MzTab(csi_file, encoding='UTF8', table_format='df')\n",
    "CSI.metadata\n",
    "df= CSI.small_molecule_table\n",
    "csifingerID= df.drop(columns= [\"calc_mass_to_charge\", \"charge\", \"taxid\", \"species\",\"database\", \"database_version\", \"spectra_ref\", \"search_engine\", \"modifications\"])\n",
    "csifingerID.to_csv(\"./results/Epemicins/structures_Epemicins.csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04f80b6f",
   "metadata": {},
   "source": [
    "### Explanation of columns\n",
    "##### mz= mass-to-charge ratio (m/z)\n",
    "##### RT= retention time (sec)\n",
    "##### intensity = intensity of the feature (AU-arbitrary units)\n",
    "##### FWHM= Full Width of the peak at Half its Maximum height\n",
    "##### num_of_masstraces\t= number of mass traces detected (single mass traces are excluded). This is relevant to the isotopic pattern\n",
    "##### isotope_distances = distance in mz between the isotopes (jumps of app. 1 is important to confirm that this is a real feature) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "db72257d-802c-4fcf-99f0-19e81e7398e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>chemical_formula</th>\n",
       "      <th>smiles</th>\n",
       "      <th>inchi_key</th>\n",
       "      <th>description</th>\n",
       "      <th>exp_mass_to_charge</th>\n",
       "      <th>retention_time</th>\n",
       "      <th>best_search_engine_score[1]</th>\n",
       "      <th>opt_global_rank</th>\n",
       "      <th>opt_global_compoundId</th>\n",
       "      <th>opt_global_compoundScanNumber</th>\n",
       "      <th>opt_global_featureId</th>\n",
       "      <th>opt_global_native_id</th>\n",
       "      <th>opt_global_adduct</th>\n",
       "      <th>opt_global_dblinks</th>\n",
       "      <th>opt_global_dbflags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14852549|14852550|25053655|24971283|53652498|5...</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCCCCCCC(CC1C(C(=O)O1)CCCCCC)OC(=O)CNC=O</td>\n",
       "      <td>ANPULBVRORHAGO</td>\n",
       "      <td>Orlistat derivative, 15g</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-167.839042</td>\n",
       "      <td>1</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>MeSH:(14852550)|PubChem:(14852549 14852550 250...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9889418|11059239</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCCCCCC1C(OC1=O)CC(CCCCCCC)OC(=O)CNC=O</td>\n",
       "      <td>RQFZMIXBLJEUGS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-168.290573</td>\n",
       "      <td>2</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>COCONUT:(CNP0260950)|KNApSAcK:(16423)|Natural ...</td>\n",
       "      <td>3178514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>71464556|71464557</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCC=CCC=CCCCCCC(CC(=O)OC(CC(=O)O)C[N+](C)(C...</td>\n",
       "      <td>WQYXCASYXUFNSI</td>\n",
       "      <td>3-hydroxylinoleylcarnitine</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-170.979093</td>\n",
       "      <td>3</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>PubChem:(71464556 71464557)|CHEBI:(73075)</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9824494</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCCC(CC1C(C(=O)O1)CCCCCCCC(C)C)OC(=O)CNC=O</td>\n",
       "      <td>SMEWLQMERHEMJK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-171.211550</td>\n",
       "      <td>4</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>COCONUT:(CNP0157254)|Natural Products:(UNPD174...</td>\n",
       "      <td>3178498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3049003</td>\n",
       "      <td>C25H45NO5</td>\n",
       "      <td>CCCCCCC(C)(CCN1C(C(=O)C(C1=O)(C)C)CCCCCCC(=O)O...</td>\n",
       "      <td>OHQKFFPFFFUNJW</td>\n",
       "      <td>Ethyl 4,4-dimethyl-3,5-dioxo-1-(3-hydroxy-3-me...</td>\n",
       "      <td>457.363953</td>\n",
       "      <td>327.701974</td>\n",
       "      <td>-250.749662</td>\n",
       "      <td>5</td>\n",
       "      <td>1571</td>\n",
       "      <td>1572</td>\n",
       "      <td>id_6254478322423518528</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=1572</td>\n",
       "      <td>[M + H4 + N]+</td>\n",
       "      <td>PubChem:(3049003)|PubChem class - safety and t...</td>\n",
       "      <td>67108866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2572</th>\n",
       "      <td>288197|2754113|7056395|7056399|11868690|118686...</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC12CCC3C(C1CCC2=O)CCC(C3(C)CC(=O)O)CC(=O)O</td>\n",
       "      <td>CEUQHWSIXGFNGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-155.792453</td>\n",
       "      <td>1</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0068730)|PubChem:(288197 2754113 7...</td>\n",
       "      <td>3153922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573</th>\n",
       "      <td>\"\"</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC(=O)C(=O)C1(CCC2C(C(=O)CCC2(C1CC(=O)O)C)(C)C)C</td>\n",
       "      <td>QWLGLRAGSFVBDI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-168.039526</td>\n",
       "      <td>2</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0136763)|Natural Products:(UNPD195...</td>\n",
       "      <td>3178496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2574</th>\n",
       "      <td>354385|494326|23425264</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC(=O)OC1CC2CC3(C(CC(C3C2(C1=C)O)(C)C)OC(=O)C)C</td>\n",
       "      <td>CLIPYJQRTQNGAM</td>\n",
       "      <td>.delta.-(sup 9,8.beta., 10.alpha.-triol, 3,8-d...</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-203.768295</td>\n",
       "      <td>3</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0216092)|Natural Products:(UNPD156...</td>\n",
       "      <td>3178498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2575</th>\n",
       "      <td>\"\"</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CC(=O)OC1CC2CC3(CC(C(C3C2(C1=C)O)(C)C)OC(=O)C)C</td>\n",
       "      <td>KEVCQSXXJWIVSR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-205.873518</td>\n",
       "      <td>4</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0261220)|Natural Products:(UNPD183...</td>\n",
       "      <td>2129920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2576</th>\n",
       "      <td>11717027</td>\n",
       "      <td>C19H28O5</td>\n",
       "      <td>CCCCCCCC(C(C(C#CC#CC(C=C)O)O)OC(=O)C)O</td>\n",
       "      <td>TYSPCPWTSAFCRA</td>\n",
       "      <td>[(8R,9R,10R)-3,8,10-trihydroxyheptadec-1-en-4,...</td>\n",
       "      <td>337.199054</td>\n",
       "      <td>465.541832</td>\n",
       "      <td>-210.247162</td>\n",
       "      <td>5</td>\n",
       "      <td>2336</td>\n",
       "      <td>2337</td>\n",
       "      <td>id_361053789030700068</td>\n",
       "      <td>controllerType=0 controllerNumber=1 scan=2337</td>\n",
       "      <td>[M + H]+</td>\n",
       "      <td>COCONUT:(CNP0186238)|Natural Products:(UNPD125...</td>\n",
       "      <td>2129922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2577 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             identifier chemical_formula  \\\n",
       "0     14852549|14852550|25053655|24971283|53652498|5...        C25H45NO5   \n",
       "1                                      9889418|11059239        C25H45NO5   \n",
       "2                                     71464556|71464557        C25H45NO5   \n",
       "3                                               9824494        C25H45NO5   \n",
       "4                                               3049003        C25H45NO5   \n",
       "...                                                 ...              ...   \n",
       "2572  288197|2754113|7056395|7056399|11868690|118686...         C19H28O5   \n",
       "2573                                                 \"\"         C19H28O5   \n",
       "2574                             354385|494326|23425264         C19H28O5   \n",
       "2575                                                 \"\"         C19H28O5   \n",
       "2576                                           11717027         C19H28O5   \n",
       "\n",
       "                                                 smiles       inchi_key  \\\n",
       "0          CCCCCCCCCCCC(CC1C(C(=O)O1)CCCCCC)OC(=O)CNC=O  ANPULBVRORHAGO   \n",
       "1            CCCCCCCCCCC1C(OC1=O)CC(CCCCCCC)OC(=O)CNC=O  RQFZMIXBLJEUGS   \n",
       "2     CCCCCC=CCC=CCCCCCC(CC(=O)OC(CC(=O)O)C[N+](C)(C...  WQYXCASYXUFNSI   \n",
       "3        CCCCCCCC(CC1C(C(=O)O1)CCCCCCCC(C)C)OC(=O)CNC=O  SMEWLQMERHEMJK   \n",
       "4     CCCCCCC(C)(CCN1C(C(=O)C(C1=O)(C)C)CCCCCCC(=O)O...  OHQKFFPFFFUNJW   \n",
       "...                                                 ...             ...   \n",
       "2572        CC12CCC3C(C1CCC2=O)CCC(C3(C)CC(=O)O)CC(=O)O  CEUQHWSIXGFNGI   \n",
       "2573   CC(=O)C(=O)C1(CCC2C(C(=O)CCC2(C1CC(=O)O)C)(C)C)C  QWLGLRAGSFVBDI   \n",
       "2574    CC(=O)OC1CC2CC3(C(CC(C3C2(C1=C)O)(C)C)OC(=O)C)C  CLIPYJQRTQNGAM   \n",
       "2575    CC(=O)OC1CC2CC3(CC(C(C3C2(C1=C)O)(C)C)OC(=O)C)C  KEVCQSXXJWIVSR   \n",
       "2576             CCCCCCCC(C(C(C#CC#CC(C=C)O)O)OC(=O)C)O  TYSPCPWTSAFCRA   \n",
       "\n",
       "                                            description  exp_mass_to_charge  \\\n",
       "0                              Orlistat derivative, 15g          457.363953   \n",
       "1                                                   NaN          457.363953   \n",
       "2                            3-hydroxylinoleylcarnitine          457.363953   \n",
       "3                                                   NaN          457.363953   \n",
       "4     Ethyl 4,4-dimethyl-3,5-dioxo-1-(3-hydroxy-3-me...          457.363953   \n",
       "...                                                 ...                 ...   \n",
       "2572                                                NaN          337.199054   \n",
       "2573                                                NaN          337.199054   \n",
       "2574  .delta.-(sup 9,8.beta., 10.alpha.-triol, 3,8-d...          337.199054   \n",
       "2575                                                NaN          337.199054   \n",
       "2576  [(8R,9R,10R)-3,8,10-trihydroxyheptadec-1-en-4,...          337.199054   \n",
       "\n",
       "      retention_time  best_search_engine_score[1]  opt_global_rank  \\\n",
       "0         327.701974                  -167.839042                1   \n",
       "1         327.701974                  -168.290573                2   \n",
       "2         327.701974                  -170.979093                3   \n",
       "3         327.701974                  -171.211550                4   \n",
       "4         327.701974                  -250.749662                5   \n",
       "...              ...                          ...              ...   \n",
       "2572      465.541832                  -155.792453                1   \n",
       "2573      465.541832                  -168.039526                2   \n",
       "2574      465.541832                  -203.768295                3   \n",
       "2575      465.541832                  -205.873518                4   \n",
       "2576      465.541832                  -210.247162                5   \n",
       "\n",
       "      opt_global_compoundId  opt_global_compoundScanNumber  \\\n",
       "0                      1571                           1572   \n",
       "1                      1571                           1572   \n",
       "2                      1571                           1572   \n",
       "3                      1571                           1572   \n",
       "4                      1571                           1572   \n",
       "...                     ...                            ...   \n",
       "2572                   2336                           2337   \n",
       "2573                   2336                           2337   \n",
       "2574                   2336                           2337   \n",
       "2575                   2336                           2337   \n",
       "2576                   2336                           2337   \n",
       "\n",
       "        opt_global_featureId                           opt_global_native_id  \\\n",
       "0     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "1     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "2     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "3     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "4     id_6254478322423518528  controllerType=0 controllerNumber=1 scan=1572   \n",
       "...                      ...                                            ...   \n",
       "2572   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2573   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2574   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2575   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "2576   id_361053789030700068  controllerType=0 controllerNumber=1 scan=2337   \n",
       "\n",
       "     opt_global_adduct                                 opt_global_dblinks  \\\n",
       "0        [M + H4 + N]+  MeSH:(14852550)|PubChem:(14852549 14852550 250...   \n",
       "1        [M + H4 + N]+  COCONUT:(CNP0260950)|KNApSAcK:(16423)|Natural ...   \n",
       "2        [M + H4 + N]+          PubChem:(71464556 71464557)|CHEBI:(73075)   \n",
       "3        [M + H4 + N]+  COCONUT:(CNP0157254)|Natural Products:(UNPD174...   \n",
       "4        [M + H4 + N]+  PubChem:(3049003)|PubChem class - safety and t...   \n",
       "...                ...                                                ...   \n",
       "2572          [M + H]+  COCONUT:(CNP0068730)|PubChem:(288197 2754113 7...   \n",
       "2573          [M + H]+  COCONUT:(CNP0136763)|Natural Products:(UNPD195...   \n",
       "2574          [M + H]+  COCONUT:(CNP0216092)|Natural Products:(UNPD156...   \n",
       "2575          [M + H]+  COCONUT:(CNP0261220)|Natural Products:(UNPD183...   \n",
       "2576          [M + H]+  COCONUT:(CNP0186238)|Natural Products:(UNPD125...   \n",
       "\n",
       "      opt_global_dbflags  \n",
       "0                     70  \n",
       "1                3178514  \n",
       "2                     34  \n",
       "3                3178498  \n",
       "4               67108866  \n",
       "...                  ...  \n",
       "2572             3153922  \n",
       "2573             3178496  \n",
       "2574             3178498  \n",
       "2575             2129920  \n",
       "2576             2129922  \n",
       "\n",
       "[2577 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example\n",
    "csifingerID= pd.read_csv(\"./results/Epemicins/structures_Epemicins.csv\",index_col= \"Unnamed: 0\")\n",
    "csifingerID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d0b7749-07f7-48fa-85c7-afa5561709b6",
   "metadata": {},
   "source": [
    "### Create a metadata csv file for GNPS from the samples.tsv file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4d147d81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>ATTRIBUTE_comment</th>\n",
       "      <th>ATTRIBUTE_MAP_ID</th>\n",
       "      <th>ATTRIBUTE_genomeID</th>\n",
       "      <th>ATTRIBUTE_media</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20210827_UMETAB219_POS_DNPM_Plate-2_MDNA_WGS_1...</td>\n",
       "      <td>MDNA_WGS_14_DNPM</td>\n",
       "      <td>MAP0</td>\n",
       "      <td>MDNA_WGS_14</td>\n",
       "      <td>DNPM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20210827_UMETAB219_POS_ISP2_Plate-2_MDNA_WGS_1...</td>\n",
       "      <td>MDNA_WGS_14_ISP2</td>\n",
       "      <td>MAP1</td>\n",
       "      <td>MDNA_WGS_14</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20210827_UMETAB219_POS_ISP2_Plate-2_MDNA_WGS_1...</td>\n",
       "      <td>MDNA_WGS_14_ISP2</td>\n",
       "      <td>MAP2</td>\n",
       "      <td>MDNA_WGS_14</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20210827_UMETAB219_POS_ISP2_Plate-2_MDNA_WGS_1...</td>\n",
       "      <td>MDNA_WGS_14_ISP2</td>\n",
       "      <td>MAP3</td>\n",
       "      <td>MDNA_WGS_14</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20210927_POS_UMETAB222_4_ISP2_Plate-10_NBC_009...</td>\n",
       "      <td>NBC_00984_ISP2</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>NBC_00984</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>20210927_POS_UMETAB222_55_ISP2_Plate-10_NBC_00...</td>\n",
       "      <td>NBC_00734_ISP2</td>\n",
       "      <td>MAP5</td>\n",
       "      <td>NBC_00734</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>20210927_POS_UMETAB222_56_DNPM_Plate-10_NBC_00...</td>\n",
       "      <td>NBC_00734_DNPM</td>\n",
       "      <td>MAP6</td>\n",
       "      <td>NBC_00734</td>\n",
       "      <td>DNPM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>20210927_POS_UMETAB222_57_FPY12_Plate-10_NBC_0...</td>\n",
       "      <td>NBC_00734_FPY12</td>\n",
       "      <td>MAP7</td>\n",
       "      <td>NBC_00734</td>\n",
       "      <td>FPY12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>20210927_POS_UMETAB222_5_DNPM_Plate-10_NBC_009...</td>\n",
       "      <td>NBC_00984_DNPM</td>\n",
       "      <td>MAP8</td>\n",
       "      <td>NBC_00984</td>\n",
       "      <td>DNPM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>20210927_POS_UMETAB222_6_FPY12_Plate-10_NBC_00...</td>\n",
       "      <td>NBC_00984_FPY12</td>\n",
       "      <td>MAP9</td>\n",
       "      <td>NBC_00984</td>\n",
       "      <td>FPY12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_010...</td>\n",
       "      <td>NBC_01014_ISP2</td>\n",
       "      <td>MAP10</td>\n",
       "      <td>NBC_01014</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21210924_UMETAB222_POS_4_ISP2_Plate-8_NBC_0052...</td>\n",
       "      <td>NBC_00520_ISP2</td>\n",
       "      <td>MAP11</td>\n",
       "      <td>NBC_00520</td>\n",
       "      <td>ISP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>21210924_UMETAB222_POS_5_DNPM_Plate-8_NBC_0052...</td>\n",
       "      <td>NBC_00520_DNPM</td>\n",
       "      <td>MAP12</td>\n",
       "      <td>NBC_00520</td>\n",
       "      <td>DNPM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>21210924_UMETAB222_POS_6_FPY12_Plate-8_NBC_005...</td>\n",
       "      <td>NBC_00520_FPY12</td>\n",
       "      <td>MAP13</td>\n",
       "      <td>NBC_00520</td>\n",
       "      <td>FPY12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             filename ATTRIBUTE_comment  \\\n",
       "0   20210827_UMETAB219_POS_DNPM_Plate-2_MDNA_WGS_1...  MDNA_WGS_14_DNPM   \n",
       "1   20210827_UMETAB219_POS_ISP2_Plate-2_MDNA_WGS_1...  MDNA_WGS_14_ISP2   \n",
       "2   20210827_UMETAB219_POS_ISP2_Plate-2_MDNA_WGS_1...  MDNA_WGS_14_ISP2   \n",
       "3   20210827_UMETAB219_POS_ISP2_Plate-2_MDNA_WGS_1...  MDNA_WGS_14_ISP2   \n",
       "4   20210927_POS_UMETAB222_4_ISP2_Plate-10_NBC_009...    NBC_00984_ISP2   \n",
       "5   20210927_POS_UMETAB222_55_ISP2_Plate-10_NBC_00...    NBC_00734_ISP2   \n",
       "6   20210927_POS_UMETAB222_56_DNPM_Plate-10_NBC_00...    NBC_00734_DNPM   \n",
       "7   20210927_POS_UMETAB222_57_FPY12_Plate-10_NBC_0...   NBC_00734_FPY12   \n",
       "8   20210927_POS_UMETAB222_5_DNPM_Plate-10_NBC_009...    NBC_00984_DNPM   \n",
       "9   20210927_POS_UMETAB222_6_FPY12_Plate-10_NBC_00...   NBC_00984_FPY12   \n",
       "10  20211006_UMETAB222_POS_7_ISP2_Plate-11_NBC_010...    NBC_01014_ISP2   \n",
       "11  21210924_UMETAB222_POS_4_ISP2_Plate-8_NBC_0052...    NBC_00520_ISP2   \n",
       "12  21210924_UMETAB222_POS_5_DNPM_Plate-8_NBC_0052...    NBC_00520_DNPM   \n",
       "13  21210924_UMETAB222_POS_6_FPY12_Plate-8_NBC_005...   NBC_00520_FPY12   \n",
       "\n",
       "   ATTRIBUTE_MAP_ID ATTRIBUTE_genomeID ATTRIBUTE_media  \n",
       "0              MAP0        MDNA_WGS_14            DNPM  \n",
       "1              MAP1        MDNA_WGS_14            ISP2  \n",
       "2              MAP2        MDNA_WGS_14            ISP2  \n",
       "3              MAP3        MDNA_WGS_14            ISP2  \n",
       "4              MAP4          NBC_00984            ISP2  \n",
       "5              MAP5          NBC_00734            ISP2  \n",
       "6              MAP6          NBC_00734            DNPM  \n",
       "7              MAP7          NBC_00734           FPY12  \n",
       "8              MAP8          NBC_00984            DNPM  \n",
       "9              MAP9          NBC_00984           FPY12  \n",
       "10            MAP10          NBC_01014            ISP2  \n",
       "11            MAP11          NBC_00520            ISP2  \n",
       "12            MAP12          NBC_00520            DNPM  \n",
       "13            MAP13          NBC_00520           FPY12  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "path= \"results/GNPSexport/interim\"\n",
    "isExist= os.path.exists(path)\n",
    "if not isExist:\n",
    "    os.mkdir(path)\n",
    "\n",
    "df= pd.read_csv(\"config/samples.tsv\", sep= \"\\t\", index_col= \"Unnamed: 0\")\n",
    "df[\"sample_name\"]=df[\"sample_name\"].replace(to_replace= r'MDNAWGS', value= 'MDNA_WGS_', regex= True)\n",
    "metadata= df.rename(columns= {\"sample_name\": \"filename\", \"comment\": \"ATTRIBUTE_comment\", \"MAPnumber\": \"ATTRIBUTE_MAPnumber\"})\n",
    "metadata[\"filename\"]= metadata[\"filename\"].astype(str) +\".mzml\"\n",
    "metadata['ATTRIBUTE_MAPnumber'] = np.arange(len(metadata))\n",
    "metadata[\"ATTRIBUTE_MAP_ID\"]= \"MAP\" + metadata[\"ATTRIBUTE_MAPnumber\"].astype(str)\n",
    "metadata= metadata.drop(columns= \"ATTRIBUTE_MAPnumber\")\n",
    "metadata['ATTRIBUTE_genomeID']=metadata['filename'].str.extract(r'(NBC_?\\d*)')\n",
    "metadata['ATTRIBUTE_genomeIDMDNA']=metadata['filename'].str.extract(r'(MDNAWGS?\\d*|MDNA_WGS_?\\d*)')\n",
    "metadata['ATTRIBUTE_genomeID']=metadata['ATTRIBUTE_genomeID'].fillna(metadata['ATTRIBUTE_genomeIDMDNA'])\n",
    "metadata[\"ATTRIBUTE_media\"]= metadata['filename'].str.extract(r'(ISP2|DNPM|FPY12\\d*)')\n",
    "metadata[\"ATTRIBUTE_comment\"]= metadata['ATTRIBUTE_genomeID'].astype(str) +\"_\" + metadata[\"ATTRIBUTE_media\"].astype(str)\n",
    "metadata=metadata.drop(columns=\"ATTRIBUTE_genomeIDMDNA\")\n",
    "#metadata['ATTRIBUTE_genomeID']= metadata['ATTRIBUTE_genomeID'].replace(to_replace= r'NBC', value= 'NBC_', regex= True)\n",
    "#metadata['ATTRIBUTE_genomeID']= metadata['ATTRIBUTE_genomeID'].replace(to_replace= r'MDNAWGS', value= 'MDNA_WGS_', regex= True)\n",
    "metadata.to_csv(\"results/GNPSexport/metadata.tsv\", sep='\\t')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60121532-63ae-41f6-a028-d130f3aa2d80",
   "metadata": {},
   "source": [
    "From the CLI, I am using the following algorithms to generate files compatible to GNPS for FBMN: \n",
    "\n",
    "#### MapAlignerPoseClustering is used to perform a linear retention time alignment, basically correct for linear shifts in retention time. \n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_MapAlignerPoseClustering.html\n",
    "\n",
    "    rule MapAlignerPoseClustering:\n",
    "        input:\n",
    "            expand(\"results/{samples}/interim/MFD_{samples}.featureXML\", samples=SAMPLES)\n",
    "        output:\n",
    "            var1= expand(\"results/Consensus/interim/MapAlignerPoseClustering_{samples}.featureXML\", samples=SAMPLES),\n",
    "            var2= expand(\"results/Consensus/interim/MapAlignerPoseClustering_{samples}.trafoXML\", samples=SAMPLES)\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/MapAlignerPoseClustering -in {input} -out {output.var1} -trafo_out {output.var2}\n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "fe932f36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Picking a reference (by size) ... done\n",
      "Progress of 'Aligning input maps':\n",
      "-- done [took 4.47 s (CPU), 4.54 s (Wall)] -- \n",
      "MapAlignerPoseClustering took 6.10 s (wall), 6.02 s (CPU), 0.30 s (system), 5.72 s (user); Peak Memory Usage: 68 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/MapAlignerPoseClustering -algorithm:max_num_peaks_considered -1 -in results/Epemicins/interim/MFD_Epemicins.featureXML results/GermicidinA/interim/MFD_GermicidinA.featureXML results/GermicidinB/interim/MFD_GermicidinB.featureXML -out results/Consensus/interim/MapAlignerPoseClustering_Epemicins.featureXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinA.featureXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinB.featureXML -trafo_out results/Consensus/interim/MapAlignerPoseClustering_Epemicins.trafoXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinA.trafoXML results/Consensus/interim/MapAlignerPoseClustering_GermicidinB.trafoXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e00257a6",
   "metadata": {},
   "source": [
    "#### Introduce the features to a protein identification file (idXML)   \n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_IDMapper.html\n",
    "\n",
    "    rule IDMapper:\n",
    "        input:\n",
    "            \"resources/emptyfile.idXML\",\n",
    "            \"results/Consensus/interim/MapAlignerPoseClustering_{samples}.featureXML\",\n",
    "            \"results/{samples}/interim/precursorcorrected_{samples}.mzML\"\n",
    "        output:\n",
    "            \"results/Consensus/interim/IDMapper_{samples}.featureXML\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/IDMapper -id {input[0]} -in {input[1]}  -spectra:in {input[2]} -out {output} \n",
    "            \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e0cbea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!resources/OpenMS-2.7.0/bin/IDMapper -id resources/emptyfile.idXML -in results/GNPSexport/interim/MapAlignerPoseClustering_Epemicins.featureXML  -spectra:in results/Epemicins/interim/precursorcorrected_Epemicins.mzML -out results/GNPSexport/interim/IDMapper_Epemicins.featureXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12202f9b",
   "metadata": {},
   "source": [
    "#### The FeatureLinkerUnlabeledKD is used to aggregate the feature information (from single files) into a ConsensusFeature, linking features from different files together, which have a smiliar m/z and rt (no MS2 data).\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FeatureLinkerUnlabeledKD.html\n",
    "\n",
    "    rule FeatureLinkerUnlabeledKD:\n",
    "        input:\n",
    "            expand(\"results/Consensus/interim/IDMapper_{samples}.featureXML\", samples=SAMPLES)\n",
    "        output:\n",
    "            \"results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/FeatureLinkerUnlabeledKD -in {input} -out {output} \n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec78e348",
   "metadata": {},
   "outputs": [],
   "source": [
    "!resources/OpenMS-2.7.0/bin/FeatureLinkerUnlabeledKD -in results/Consensus/interim/IDMapper_Epemicins.featureXML results/Consensus/interim/IDMapper_GermicidinA.featureXML results/Consensus/interim/IDMapper_GermicidinB.featureXML -out results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670a0817",
   "metadata": {},
   "source": [
    "#### Filter out the features that do not have an MS2 pattern\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_FileFilter.html\n",
    "\n",
    "    rule FileFilter:\n",
    "        input:\n",
    "            \"results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML\"\n",
    "        output:\n",
    "            \"results/Consensus/filtered.consensusXML\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/FileFilter -id:remove_unannotated_features -in {input} -out {output} \n",
    "            \"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f85a472d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!resources/OpenMS-2.7.0/bin/FileFilter -id:remove_unannotated_features -in results/Consensus/interim/FeatureLinkerUnlabeledKD.consensusXML -out results/Consensus/filtered.consensusXML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04145cbe",
   "metadata": {},
   "source": [
    "\n",
    "#### GNPS_export creates an mgf file with only the MS2 information of all files (introduce mzml files with spaces between them)\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_GNPSExport.html\n",
    "\n",
    "    rule GNPS_export:\n",
    "        input:\n",
    "            \"results/Consensus/filtered.consensusXML\",\n",
    "            expand(\"results/{samples}/interim/precursorcorrected_{samples}.mzML\", samples=SAMPLES)\n",
    "        output:\n",
    "            \"results/GNPSexport/MSMS.mgf\" \n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/GNPSExport -in_cm {input[0]} -in_mzml {input[1]} -out {output} \n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d29ee426",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress of 'parsing features and ms2 identifications...':\n",
      "  1.63 %               GNPSExport took 2.62 s (wall), 2.58 s (CPU), 0.16 s (system), 2.42 s (user); Peak Memory Usage: 87 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/GNPSExport -ini resources/GNPSExport.ini -in_cm results/Consensus/filtered.consensusXML -in_mzml results/Epemicins/interim/precursorcorrected_Epemicins.mzML results/GermicidinA/interim/precursorcorrected_GermicidinA.mzML results/GermicidinB/interim/precursorcorrected_GermicidinB.mzML -out results/GNPSexport/MSMS.mgf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ac98624",
   "metadata": {},
   "source": [
    "#### export the consensusXML file to a txt file for GNPS\n",
    "Documentation: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/release/latest/html/TOPP_TextExporter.html\n",
    "    \n",
    "    rule txt_export:\n",
    "        input:\n",
    "            \"results/Consensus/filtered.consensusXML\"\n",
    "        output:\n",
    "            \"results/GNPSexport/FeatureQuantificationTable.txt\" \n",
    "        shell:\n",
    "            \"\"\"\n",
    "            resources/OpenMS-2.7.0/bin/TextExporter -in {input} -out {output}\n",
    "            \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bac296b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TextExporter took 0.09 s (wall), 0.09 s (CPU), 0.01 s (system), 0.08 s (user); Peak Memory Usage: 19 MB.\n"
     ]
    }
   ],
   "source": [
    "!resources/OpenMS-2.7.0/bin/TextExporter -in results/Consensus/filtered.consensusXML -out results/GNPSexport/FeatureQuantificationTable.txt"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "edde62aa2661007f0756e9790e7a328c288a583bf6ce768a355147dac67c8db8"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
